

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Python API Reference &mdash; pyFuRNAce 1.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/copybutton.css?v=76b2166b" />


      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=8d563738"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script src="_static/clipboard.min.js?v=a7894cd8"></script>
      <script src="_static/copybutton.js?v=f281be69"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Limitations &amp; Workarounds" href="limitations.html" />
    <link rel="prev" title="Install and run" href="installation.html" />
</head>

<body class="wy-body-for-nav">
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >



          <a href="index.html" class="icon icon-home">
            pyFuRNAce
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="about.html">About pyFuRNAce</a><ul>
<li class="toctree-l2"><a class="reference internal" href="about.html#web-application">Web Application</a><ul>
<li class="toctree-l3"><a class="reference internal" href="about.html#modules">Modules</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="about.html#python-api">Python API</a></li>
<li class="toctree-l2"><a class="reference internal" href="about.html#citation">Citation</a></li>
<li class="toctree-l2"><a class="reference internal" href="about.html#license">License</a></li>
<li class="toctree-l2"><a class="reference internal" href="about.html#acknowledgements">Acknowledgements</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Install and run</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Python API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#what-s-the-idea">What’s the idea?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#design-core-objects">Design Core Objects</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.core.strand">Strand</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.strand.Strand"><code class="docutils literal notranslate"><span class="pre">Strand</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.strand.StrandsBlock"><code class="docutils literal notranslate"><span class="pre">StrandsBlock</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#motif">Motif</a></li>
<li class="toctree-l3"><a class="reference internal" href="#origami">Origami</a></li>
<li class="toctree-l3"><a class="reference internal" href="#sequence">Sequence</a></li>
<li class="toctree-l3"><a class="reference internal" href="#basepair">Basepair</a></li>
<li class="toctree-l3"><a class="reference internal" href="#common-variables">Common Variables</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#useful-collections">Useful collections</a></li>
<li class="toctree-l4"><a class="reference internal" href="#translators-for-strings-characters">Translators for strings characters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.core.symbols">Common functions</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.AmbiguosStructure"><code class="docutils literal notranslate"><span class="pre">AmbiguosStructure</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.MotifStructureError"><code class="docutils literal notranslate"><span class="pre">MotifStructureError</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.base_pair_difference"><code class="docutils literal notranslate"><span class="pre">base_pair_difference()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.base_pair_distance"><code class="docutils literal notranslate"><span class="pre">base_pair_distance()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.complement"><code class="docutils literal notranslate"><span class="pre">complement()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.dot_bracket_to_pair_map"><code class="docutils literal notranslate"><span class="pre">dot_bracket_to_pair_map()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.dot_bracket_to_stacks"><code class="docutils literal notranslate"><span class="pre">dot_bracket_to_stacks()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.dot_bracket_to_tree"><code class="docutils literal notranslate"><span class="pre">dot_bracket_to_tree()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.folding_barriers"><code class="docutils literal notranslate"><span class="pre">folding_barriers()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.gc_content"><code class="docutils literal notranslate"><span class="pre">gc_content()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.hamming_distance"><code class="docutils literal notranslate"><span class="pre">hamming_distance()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.mutate_nucleotide"><code class="docutils literal notranslate"><span class="pre">mutate_nucleotide()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.pair_map_to_dot_bracket"><code class="docutils literal notranslate"><span class="pre">pair_map_to_dot_bracket()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.pair_nucleotide"><code class="docutils literal notranslate"><span class="pre">pair_nucleotide()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.reverse_complement"><code class="docutils literal notranslate"><span class="pre">reverse_complement()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.rotate_dot_bracket"><code class="docutils literal notranslate"><span class="pre">rotate_dot_bracket()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.core.symbols.tree_to_dot_bracket"><code class="docutils literal notranslate"><span class="pre">tree_to_dot_bracket()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#motifs">Motifs</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.motifs.stem">Stem</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.stem.Stem"><code class="docutils literal notranslate"><span class="pre">Stem</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.motifs.dovetail">Dovetail</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.dovetail.Dovetail"><code class="docutils literal notranslate"><span class="pre">Dovetail</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.motifs.loops">Loops</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.loops.Loop"><code class="docutils literal notranslate"><span class="pre">Loop</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.loops.TetraLoop"><code class="docutils literal notranslate"><span class="pre">TetraLoop</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.motifs.kissing_loops">Kissing loops</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.BranchedDimer"><code class="docutils literal notranslate"><span class="pre">BranchedDimer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop"><code class="docutils literal notranslate"><span class="pre">BranchedKissingLoop</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer"><code class="docutils literal notranslate"><span class="pre">KissingDimer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer120"><code class="docutils literal notranslate"><span class="pre">KissingDimer120</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop"><code class="docutils literal notranslate"><span class="pre">KissingLoop</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop120"><code class="docutils literal notranslate"><span class="pre">KissingLoop120</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop180"><code class="docutils literal notranslate"><span class="pre">KissingLoop180</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.motifs.aptamers">Aptamers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Aptamer"><code class="docutils literal notranslate"><span class="pre">Aptamer</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Biotin"><code class="docutils literal notranslate"><span class="pre">Biotin()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Broccoli"><code class="docutils literal notranslate"><span class="pre">Broccoli()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Ispinach"><code class="docutils literal notranslate"><span class="pre">Ispinach()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.L7Ae"><code class="docutils literal notranslate"><span class="pre">L7Ae()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.MS2"><code class="docutils literal notranslate"><span class="pre">MS2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.MalachiteGreen"><code class="docutils literal notranslate"><span class="pre">MalachiteGreen()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.MalachiteGreenShort"><code class="docutils literal notranslate"><span class="pre">MalachiteGreenShort()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Mango"><code class="docutils literal notranslate"><span class="pre">Mango()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.PIP3"><code class="docutils literal notranslate"><span class="pre">PIP3()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.PIP3_mut1"><code class="docutils literal notranslate"><span class="pre">PIP3_mut1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.PIP3_mut3"><code class="docutils literal notranslate"><span class="pre">PIP3_mut3()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.PIP3_mut5"><code class="docutils literal notranslate"><span class="pre">PIP3_mut5()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.PP7"><code class="docutils literal notranslate"><span class="pre">PP7()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Pepper"><code class="docutils literal notranslate"><span class="pre">Pepper()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Streptavidin"><code class="docutils literal notranslate"><span class="pre">Streptavidin()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.TAR_TAT"><code class="docutils literal notranslate"><span class="pre">TAR_TAT()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Thrombin_exosite1"><code class="docutils literal notranslate"><span class="pre">Thrombin_exosite1()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.Thrombin_exosite2"><code class="docutils literal notranslate"><span class="pre">Thrombin_exosite2()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.motifs.aptamers.create_aptamer"><code class="docutils literal notranslate"><span class="pre">create_aptamer()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#utilities">Utilities</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.utils.motif_lib">Motif Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.Utils"><code class="docutils literal notranslate"><span class="pre">Utils</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.start_end_stem"><code class="docutils literal notranslate"><span class="pre">start_end_stem()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.stem_cap_link"><code class="docutils literal notranslate"><span class="pre">stem_cap_link()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.vertical_double_link"><code class="docutils literal notranslate"><span class="pre">vertical_double_link()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.vertical_link"><code class="docutils literal notranslate"><span class="pre">vertical_link()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.design.utils.origami_lib">Origami Utilities</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.convert_angles_to_dt"><code class="docutils literal notranslate"><span class="pre">convert_angles_to_dt()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.ipython_clickable_txt"><code class="docutils literal notranslate"><span class="pre">ipython_clickable_txt()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.ipython_display_3D"><code class="docutils literal notranslate"><span class="pre">ipython_display_3D()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.ipython_display_txt"><code class="docutils literal notranslate"><span class="pre">ipython_display_txt()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.simple_origami"><code class="docutils literal notranslate"><span class="pre">simple_origami()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_2_helix"><code class="docutils literal notranslate"><span class="pre">template_2_helix()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_3_arms_droplet"><code class="docutils literal notranslate"><span class="pre">template_3_arms_droplet()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_4_arms_droplet"><code class="docutils literal notranslate"><span class="pre">template_4_arms_droplet()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_pentagon_tile"><code class="docutils literal notranslate"><span class="pre">template_pentagon_tile()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_rectangle_10H_3X"><code class="docutils literal notranslate"><span class="pre">template_rectangle_10H_3X()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_rna_filament"><code class="docutils literal notranslate"><span class="pre">template_rna_filament()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.design.utils.origami_lib.template_rna_filament_ispinach"><code class="docutils literal notranslate"><span class="pre">template_rna_filament_ispinach()</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#module-pyfurnace.generate.road">Sequence Generation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#pyfurnace.generate.road.generate_road"><code class="docutils literal notranslate"><span class="pre">generate_road()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#pyfurnace.generate.road.parallel_road"><code class="docutils literal notranslate"><span class="pre">parallel_road()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#prepare">Prepare</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.prepare.utils">Primers</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.prepare.utils.annealing_temp"><code class="docutils literal notranslate"><span class="pre">annealing_temp()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.prepare.utils.auto_design_primers"><code class="docutils literal notranslate"><span class="pre">auto_design_primers()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.prepare.utils.calculate_gc"><code class="docutils literal notranslate"><span class="pre">calculate_gc()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.prepare.utils.check_dimer"><code class="docutils literal notranslate"><span class="pre">check_dimer()</span></code></a></li>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.prepare.utils.make_tm_calculator"><code class="docutils literal notranslate"><span class="pre">make_tm_calculator()</span></code></a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#module-pyfurnace.prepare.oxdna_sim">Oxdna simulation setup</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#pyfurnace.prepare.oxdna_sim.oxdna_simulations"><code class="docutils literal notranslate"><span class="pre">oxdna_simulations()</span></code></a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="limitations.html">Limitations &amp; Workarounds</a><ul>
<li class="toctree-l2"><a class="reference internal" href="limitations.html#pyfurnace-folding-barrier-estimation">PyFuRNAce folding barrier estimation</a></li>
<li class="toctree-l2"><a class="reference internal" href="limitations.html#revolvr">Revolvr</a></li>
<li class="toctree-l2"><a class="reference internal" href="limitations.html#viennarna">ViennaRNA</a></li>
<li class="toctree-l2"><a class="reference internal" href="limitations.html#oxdna-analysis-tools">oxDNA Analysis Tools</a></li>
<li class="toctree-l2"><a class="reference internal" href="limitations.html#references">References</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="gui_examples.html">Graphical User Interface Examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="gui_examples.html#including-a-custom-aptamer">Including a custom aptamer</a><ul>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#import-the-trna-aptamer-sequence">1) Import the tRNA-aptamer sequence</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#isolate-the-aptamer-motif">2) Isolate the aptamer motif</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#optional-add-3d-structure">3) Optional: Add 3D structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#python-code">4) Python code</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="gui_examples.html#custom-motif-origami-with-single-stranded-region">Custom motif: origami with single-stranded region</a><ul>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#create-an-origami-with-two-stems">1) Create an origami with two stems</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#custom-motif-creation">2) Custom motif creation</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#complete-the-structure">3) Complete the structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#equivalent-code">4) Equivalent Code</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="gui_examples.html#adding-an-overhang-at-the-5-or-3-end">Adding an overhang at the 5’ or 3’ end</a><ul>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#create-an-origami">1) Create an origami</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#add-a-line-and-a-connection">2) Add a line and a connection</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#create-and-add-the-single-stranded-motif">3) Create and add the single-stranded motif</a></li>
<li class="toctree-l4"><a class="reference internal" href="gui_examples.html#connect-the-single-stranded-motif">4) Connect the single-stranded motif</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="script_examples.html">Python Scripting Examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="script_examples.html#two-helices-with-flaps-and-optimization">Two helices with FLAPs (and optimization)</a></li>
<li class="toctree-l3"><a class="reference internal" href="script_examples.html#rna-filament">RNA filament</a></li>
<li class="toctree-l3"><a class="reference internal" href="script_examples.html#folding-barrier-optimization">Folding barrier optimization</a><ul>
<li class="toctree-l4"><a class="reference internal" href="script_examples.html#full-optimisation">Full optimisation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="script_examples.html#jupyter-notebooks">Jupyter Notebooks</a><ul>
<li class="toctree-l3"><a class="reference internal" href="notebooks/droplets.html">RNA droplets</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/droplets.html#Droplet-A-with-Malachite-Green-aptamer">Droplet A with Malachite Green aptamer</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/droplets.html#Droplet-B-with-Broccoli-aptamer">Droplet B with Broccoli aptamer</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/rna_filament.html">RNA filaments assembly tutorial</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/rna_filament.html#Create-a-basic-RNA-structure">Create a basic RNA structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/rna_filament.html#Adjust-the-ends-to-make-a-tile">Adjust the ends to make a tile</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/rna_filament.html#Repeat-the-tiles-to-make-a-filament-section">Repeat the tiles to make a filament section</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/rna_filament.html#Repeat-the-sections-to-make-a-filament">Repeat the sections to make a filament</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/rna_filament.html#Generete-the-sequence-for-a-tile">Generete the sequence for a tile</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="notebooks/ROAD_origami.html">A collection of origami from ROAD</a><ul>
<li class="toctree-l4"><a class="reference internal" href="notebooks/ROAD_origami.html#3H-2DT">3H-2DT</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/ROAD_origami.html#3H-3DT">3H-3DT</a></li>
<li class="toctree-l4"><a class="reference internal" href="notebooks/ROAD_origami.html#3H-4DT">3H-4DT</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyFuRNAce</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Python API Reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">

  <section id="python-api-reference">
<span id="api"></span><h1>Python API Reference<a class="headerlink" href="#python-api-reference" title="Link to this heading"></a></h1>
<section id="what-s-the-idea">
<h2>What’s the idea?<a class="headerlink" href="#what-s-the-idea" title="Link to this heading"></a></h2>
<p>PyFuRNAce is inspired by the ROAD RNA origami software (Geary et al., Nat. Chem., 2021), which uses text-based RNA diagrams for RNA nanostructures design.</p>
<p>To facilitate the design of RNA nanostructures, pyFuRNAce aims to provide a simple and flexible Python package for creating, manipulating, and visualizing RNA with text-based diagrams.</p>
<p>The approach of pyFuRNAce is to hierarchically build RNA nanostructures from simple building blocks: strands and motifs.</p>
<p>Therefore, there are three main classes:</p>
<ol class="arabic">
<li><p><strong>Strand</strong>
A strand is simply a string of ASCII characters, which contains nucleotides (e.g., AUCG) and/or characters to route the strands on the page (e.g., <code class="docutils literal notranslate"><span class="pre">-/\|</span></code>). If we imagine drawing a strand on a page, each symbol has an <code class="docutils literal notranslate"><span class="pre">(x,</span> <span class="pre">y)</span></code> position. In pyFuRNAce, each character position is calculated automatically, as the strand is defined by:</p>
<ul class="simple">
<li><p>strand characters (e.g., <code class="docutils literal notranslate"><span class="pre">A-UCG|CC—</span></code>)</p></li>
<li><p>start position (e.g., <code class="docutils literal notranslate"><span class="pre">(0,</span> <span class="pre">0)</span></code>)</p></li>
<li><p>start direction (e.g., <code class="docutils literal notranslate"><span class="pre">pf.RIGHT</span></code>)</p></li>
<li><p>directionality of the sequence (e.g., <code class="docutils literal notranslate"><span class="pre">'53'</span></code>; optional)</p></li>
</ul>
<p>A <code class="docutils literal notranslate"><span class="pre">Strand</span></code> can also have 3D coordinates associated with the sequence, allowing the assembly of 3D models in real time.</p>
</li>
<li><p><strong>Motif</strong>
The motif is the building block of the RNA origami, exploiting the vast library of RNA aptamers. In pyFuRNAce, a <code class="docutils literal notranslate"><span class="pre">Motif</span></code> is simply a collection of strands which can have base pairs. The base pairs define the secondary structure of the motifs. Internally, base pairs are represented as strand–position pairs in a dictionary. In the Python API, the secondary structure can be set and obtained in the more common <a class="reference external" href="https://www.tbi.univie.ac.at/RNA/ViennaRNA/doc/html/io/rna_structures.html">dot-bracket notation</a>. In summary, a motif is made of:</p>
<ul class="simple">
<li><p>strands (e.g., <code class="docutils literal notranslate"><span class="pre">[Strand('CCAGG'),</span> <span class="pre">Strand('CCGGG',</span> <span class="pre">start=(4,</span> <span class="pre">0),</span> <span class="pre">direction=pf.LEFT)]</span></code>)</p></li>
<li><p>structure (e.g., <code class="docutils literal notranslate"><span class="pre">&quot;((.((&amp;)).))&quot;</span></code>)</p></li>
</ul>
<p>A large library of pre-defined motifs and aptamers is provided in pyFuRNAce.</p>
</li>
<li><p><strong>Origami</strong>
The origami is the final RNA nanostructure. In pyFuRNAce, an <code class="docutils literal notranslate"><span class="pre">Origami</span></code> is simply a collection of motifs in a grid. The <code class="docutils literal notranslate"><span class="pre">Origami</span></code> concatenates motifs horizontally (in rows) and vertically, automatically joining the vertical junctions. In principle, any RNA nanostructure can be represented as an origami:</p>
<ul class="simple">
<li><p>motifs (e.g., <code class="docutils literal notranslate"><span class="pre">[[pf.TetraLoop(),</span> <span class="pre">pf.Stem(10),</span> <span class="pre">pf.Dovetail(-1)],</span> <span class="pre">[pf.KissingLoop180(),</span> <span class="pre">pf.Stem(8),</span> <span class="pre">pf.Dovetail(-2)]]</span></code>)</p></li>
</ul>
</li>
</ol>
<p>When your nanostructure blueprint is ready, pyFuRNAce provides functions to optimize the RNA sequence (in the <code class="docutils literal notranslate"><span class="pre">generate</span></code> subpackage), design PCR primers, or set up oxDNA simulations (<code class="docutils literal notranslate"><span class="pre">prepare</span></code> subpackage).</p>
<p>Examples of how to use the pyFuRNAce API are available in the <a class="reference internal" href="tutorials.html#tutorials"><span class="std std-ref">Tutorials</span></a> section.</p>
</section>
<section id="design-core-objects">
<h2>Design Core Objects<a class="headerlink" href="#design-core-objects" title="Link to this heading"></a></h2>
<section id="module-pyfurnace.design.core.strand">
<span id="strand"></span><h3>Strand<a class="headerlink" href="#module-pyfurnace.design.core.strand" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.strand.</span></span><span class="sig-name descname"><span class="pre">Strand</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directionality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'53'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strands_block</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Callback</span></code></p>
<p>Represents a single RNA strand in a hybrid 1D/2D/3D representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>strand</strong><span class="classifier">str, optional</span></dt><dd><p>The strand text characters (default is empty).</p>
</dd>
<dt><strong>directionality</strong><span class="classifier">{‘53’, ‘35’}, optional</span></dt><dd><p>The directionality of the strand (default is ‘53’).</p>
</dd>
<dt><strong>start</strong><span class="classifier">Position, optional</span></dt><dd><p>The starting position of the strand in 2D space (default is Position.zero()).
If set to ‘minimal’, the start position is calculated as the minimal
start position required to draw the strand in 2D to avoid negative coordinates.</p>
</dd>
<dt><strong>direction</strong><span class="classifier">Direction, optional</span></dt><dd><p>The initial direction of the strand in 2D space (default is Direction.RIGHT).</p>
</dd>
<dt><strong>coords</strong><span class="classifier">Coords, optional</span></dt><dd><p>The 3D coordinates for the strand (if not provided, they will be calculated).</p>
</dd>
<dt><strong>strands_block</strong><span class="classifier">StrandsBlock, optional</span></dt><dd><p>The block of strands to which the strand belongs.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Arbitrary keyword arguments passed to the Callback class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.coords" title="pyfurnace.design.core.strand.Strand.coords"><code class="xref py py-obj docutils literal notranslate"><span class="pre">coords</span></code></a><span class="classifier">Coords</span></dt><dd><p>The oxDNA 3D coordinates of the strand.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.direction" title="pyfurnace.design.core.strand.Strand.direction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">direction</span></code></a><span class="classifier">Direction</span></dt><dd><p>The starting (x, y) 2D direction to build the strand in 2D.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.directionality" title="pyfurnace.design.core.strand.Strand.directionality"><code class="xref py py-obj docutils literal notranslate"><span class="pre">directionality</span></code></a><span class="classifier">str</span></dt><dd><p>The directionality of the sequence (‘53’ or ‘35’)</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.directions" title="pyfurnace.design.core.strand.Strand.directions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">directions</span></code></a><span class="classifier">Tuple[Direction]</span></dt><dd><p>A tuple of the directions of each character of the strand in 2D space (x,y coordinates).</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.end" title="pyfurnace.design.core.strand.Strand.end"><code class="xref py py-obj docutils literal notranslate"><span class="pre">end</span></code></a><span class="classifier">Position</span></dt><dd><p>The last position of the strand: it is the last element of the positions.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.end_direction" title="pyfurnace.design.core.strand.Strand.end_direction"><code class="xref py py-obj docutils literal notranslate"><span class="pre">end_direction</span></code></a><span class="classifier">Direction</span></dt><dd><p>The last direction of the strand.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.max_pos" title="pyfurnace.design.core.strand.Strand.max_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></a><span class="classifier">Position</span></dt><dd><p>The maximum x, y coordinates of the strand in 2D space.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.min_pos" title="pyfurnace.design.core.strand.Strand.min_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></a><span class="classifier">Position</span></dt><dd><p>The minimum x, y coordinates of the strand in 2D space.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.minimal_dimensions" title="pyfurnace.design.core.strand.Strand.minimal_dimensions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">minimal_dimensions</span></code></a><span class="classifier">Position</span></dt><dd><p>The minimum start Positions required to draw the strand in 2D without reaching negative positions.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.next_pos" title="pyfurnace.design.core.strand.Strand.next_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">next_pos</span></code></a><span class="classifier">Position</span></dt><dd><p>The next position of the strand: it is the last position plus the end_direction.</p>
</dd>
<dt><strong>pk_info</strong><span class="classifier">Optional[Dict]</span></dt><dd><p>The pseudoknot information of the strand (if any).</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.positions" title="pyfurnace.design.core.strand.Strand.positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></a><span class="classifier">Tuple[Position]</span></dt><dd><p>A tuple of the positions of each character of the strand in 2D space (x,y coordinates).</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.prec_pos" title="pyfurnace.design.core.strand.Strand.prec_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">prec_pos</span></code></a><span class="classifier">Position</span></dt><dd><p>The preceding position of the strand: it is the start position minus the direction.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.seq_positions" title="pyfurnace.design.core.strand.Strand.seq_positions"><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></a><span class="classifier">Tuple[Position]</span></dt><dd><p>The positions of each nucleotide in the strand sequence (x,y coordinates).</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.sequence" title="pyfurnace.design.core.strand.Strand.sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></a><span class="classifier">Sequence</span></dt><dd><p>The complete sequence of the strand.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.sequence_list" title="pyfurnace.design.core.strand.Strand.sequence_list"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence_list</span></code></a><span class="classifier">List[Sequence]</span></dt><dd><p>The list of consecutive sequences in the strand.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.sequence_slice" title="pyfurnace.design.core.strand.Strand.sequence_slice"><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence_slice</span></code></a><span class="classifier">List[slice]</span></dt><dd><p>The list of slices of the sequence in the strand.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.start" title="pyfurnace.design.core.strand.Strand.start"><code class="xref py py-obj docutils literal notranslate"><span class="pre">start</span></code></a><span class="classifier">Position</span></dt><dd><p>Start position of the strand in the 2D representation.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.strand" title="pyfurnace.design.core.strand.Strand.strand"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strand</span></code></a><span class="classifier">str</span></dt><dd><p>The strand text characters.</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.strands_block" title="pyfurnace.design.core.strand.Strand.strands_block"><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands_block</span></code></a><span class="classifier">StrandsBlock</span></dt><dd><p>The block of strands that has locked 3D coordinates transformation (their relative position in the 3D space doesn’t change when joined to other strands).</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.copy" title="pyfurnace.design.core.strand.Strand.copy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></a>([callback])</p></td>
<td><p>Create a copy of the current strand, including coordinates and attributes.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.draw" title="pyfurnace.design.core.strand.Strand.draw"><code class="xref py py-obj docutils literal notranslate"><span class="pre">draw</span></code></a>([canvas, return_string, plane])</p></td>
<td><p>Draw the strand in 2D space using a canvas.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.extend" title="pyfurnace.design.core.strand.Strand.extend"><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend</span></code></a>(direction[, until, check])</p></td>
<td><p>Extend the strand in the specified direction by a given length.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.flip" title="pyfurnace.design.core.strand.Strand.flip"><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code></a>([horizontally, vertically, flip_start])</p></td>
<td><p>Flip the 2D structure of the strand.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.get_base_at_pos" title="pyfurnace.design.core.strand.Strand.get_base_at_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_base_at_pos</span></code></a>(pos)</p></td>
<td><p>Get the base at the specified 2D position in the strand.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.get_char_at_pos" title="pyfurnace.design.core.strand.Strand.get_char_at_pos"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_char_at_pos</span></code></a>(pos)</p></td>
<td><p>Get the character at the specified 2D position in the strand.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.get_position_map" title="pyfurnace.design.core.strand.Strand.get_position_map"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code></a>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.insert" title="pyfurnace.design.core.strand.Strand.insert"><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code></a>(idx, val)</p></td>
<td><p>Insert a character into the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.invert" title="pyfurnace.design.core.strand.Strand.invert"><code class="xref py py-obj docutils literal notranslate"><span class="pre">invert</span></code></a>()</p></td>
<td><p>Invert the start and end positions of the strand, effectively reversing the sequence directionalty and 3D coordinates orientation.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.join" title="pyfurnace.design.core.strand.Strand.join"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join</span></code></a>(other[, trigger_callbacks])</p></td>
<td><p>Join this strand with another strand, aligning their start and end positions and direction.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.join_strands" title="pyfurnace.design.core.strand.Strand.join_strands"><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code></a>(strand1, strand2)</p></td>
<td><p>Join two strands together if they match in the right direction and have a compatible directionality.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.pop" title="pyfurnace.design.core.strand.Strand.pop"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code></a>(idx)</p></td>
<td><p>Remove and return a character from the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.reverse" title="pyfurnace.design.core.strand.Strand.reverse"><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse</span></code></a>()</p></td>
<td><p>Reverse the directionality of the strand sequence (5' to 3' becomes 3' to 5' and vice versa).</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.save_3d_model" title="pyfurnace.design.core.strand.Strand.save_3d_model"><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code></a>([filename, return_text, pdb])</p></td>
<td><p>Save the 3D structure of the strand in oxDNA format (.dat, .top) and optionally as PDB.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.shift" title="pyfurnace.design.core.strand.Strand.shift"><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code></a>(shift_position[, check])</p></td>
<td><p>Shift the strand position in the 2D layout.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.Strand.transform" title="pyfurnace.design.core.strand.Strand.transform"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transform</span></code></a>(transformation_matrix)</p></td>
<td><p>Apply a transformation matrix to the strand's 3D coordinates.</p></td>
</tr>
</tbody>
</table>
<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.coords">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">coords</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Coords</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.coords" title="Link to this definition"></a></dt>
<dd><p>The oxDNA 3D coordinates of the strand.
If the coordinates are not set, they are calculated for a strand in
a perfect helix with the same length of the strand.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.copy">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">callback</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.copy" title="Link to this definition"></a></dt>
<dd><p>Create a copy of the current strand, including coordinates and attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>callback</strong><span class="classifier">callable, default None</span></dt><dd><p>A callback function to be registered in the copied strand.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Strand</dt><dd><p>A deep copy of the strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.direction">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">direction</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Direction</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.direction" title="Link to this definition"></a></dt>
<dd><p>The starting (x, y) 2D direction to build the strand in 2D.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.directionality">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">directionality</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'53'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'35'</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.directionality" title="Link to this definition"></a></dt>
<dd><p>The directionality of the sequence (‘53’ or ‘35’)</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.directions">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">directions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Direction</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.directions" title="Link to this definition"></a></dt>
<dd><p>A tuple of the directions of each character of the strand in 2D space
(x,y coordinates).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.draw">
<span class="sig-name descname"><span class="pre">draw</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">canvas</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_string</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">plane</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.draw"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.draw" title="Link to this definition"></a></dt>
<dd><p>Draw the strand in 2D space using a canvas.
The canvas is a list of strings, where each string is a line of the canvas.
The strand is drawn using the characters and positions of the strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>canvas</strong><span class="classifier">list, default None</span></dt><dd><p>The canvas to draw the strand on. If None, a new canvas is created.</p>
</dd>
<dt><strong>return_string</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, return the drawn canvas as a string (for printing).</p>
</dd>
<dt><strong>plane</strong><span class="classifier">int, default 0</span></dt><dd><p>The plane to draw the strand on (experimental). If 0, the strand
is drawn in the default plane.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str or None</dt><dd><p>The drawn canvas as a string if return_string is True, otherwise None.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>MotifStructureError</dt><dd><p>If the canvas is not large enough to draw the strand or a position is
trying to overwrite a symbol in the canvas.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.end">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Position</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.end" title="Link to this definition"></a></dt>
<dd><p>The last position of the strand: it is the last element of the positions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.end_direction">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">end_direction</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Direction</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.end_direction" title="Link to this definition"></a></dt>
<dd><p>The last direction of the strand.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.extend">
<span class="sig-name descname"><span class="pre">extend</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">direction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">until</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.extend"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.extend" title="Link to this definition"></a></dt>
<dd><p>Extend the strand in the specified direction by a given length.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>direction</strong><span class="classifier">Union[Position, Tuple[int, int]], default (1, 0)</span></dt><dd><p>The direction to extend the strand in. It can be a Position object or
a tuple of (x, y) coordinates. It must contain coordinates that are either
1, -1 or 0, where 0 is the coordinate that does not change.</p>
</dd>
<dt><strong>until</strong><span class="classifier">Union[Position, Tuple[int, int]], default None</span></dt><dd><p>The position until which to extend the strand. If None, the strand is
extended until the origin position (0, 0).</p>
</dd>
<dt><strong>check</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, check the direction and until position for validity.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.flip">
<span class="sig-name descname"><span class="pre">flip</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">horizontally</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertically</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">flip_start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.flip"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.flip" title="Link to this definition"></a></dt>
<dd><p>Flip the 2D structure of the strand. This affects the orientation on the
canvas but does not revert the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>horizontally</strong><span class="classifier">bool, default=True</span></dt><dd><p>Whether to flip the strand horizontally (mirror along the y-axis).</p>
</dd>
<dt><strong>vertically</strong><span class="classifier">bool, default=False</span></dt><dd><p>Whether to flip the strand vertically (mirror along the x-axis).</p>
</dd>
<dt><strong>flip_start</strong><span class="classifier">bool, default=True</span></dt><dd><p>Whether to also flip the starting point of the strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.get_base_at_pos">
<span class="sig-name descname"><span class="pre">get_base_at_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.get_base_at_pos"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.get_base_at_pos" title="Link to this definition"></a></dt>
<dd><p>Get the base at the specified 2D position in the strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pos</strong><span class="classifier">Position or tuple of int</span></dt><dd><p>The position to get the base from.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The base at the specified position.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.get_char_at_pos">
<span class="sig-name descname"><span class="pre">get_char_at_pos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pos</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.get_char_at_pos"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.get_char_at_pos" title="Link to this definition"></a></dt>
<dd><p>Get the character at the specified 2D position in the strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pos</strong><span class="classifier">Position or tuple of int</span></dt><dd><p>The position to get the character from.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The character at the specified position.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.get_position_map">
<span class="sig-name descname"><span class="pre">get_position_map</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.get_position_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.get_position_map" title="Link to this definition"></a></dt>
<dd><p>Get a dictionary of positions as keys and the corresponding
characters as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Dict[Tuple[int, int], str]</dt><dd><p>A dictionary mapping positions to characters.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.insert">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">val</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.insert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.insert" title="Link to this definition"></a></dt>
<dd><p>Insert a character into the strand at the specified index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>idx</strong><span class="classifier">int</span></dt><dd><p>The index at which to insert the character.</p>
</dd>
<dt><strong>val</strong><span class="classifier">str</span></dt><dd><p>The character(s) to insert into the strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.invert">
<span class="sig-name descname"><span class="pre">invert</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.invert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.invert" title="Link to this definition"></a></dt>
<dd><p>Invert the start and end positions of the strand,
effectively reversing the sequence directionalty
and 3D coordinates orientation.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Strand</dt><dd><p>The updated strand instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.join">
<span class="sig-name descname"><span class="pre">join</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trigger_callbacks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.join"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.join" title="Link to this definition"></a></dt>
<dd><p>Join this strand with another strand, aligning their start and end
positions and direction. This behaves like the &#64;staticmethod join_strands,
but it modifies the current strand instead of creating a new one.
Check the join_strands method for more details.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>other</strong><span class="classifier">Strand</span></dt><dd><p>The other strand to join.</p>
</dd>
<dt><strong>trigger_callbacks</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, trigger the callbacks after joining the strands.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Strand or None</dt><dd><p>The joined strand, or None if the strands cannot be joined.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.join_strands">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">join_strands</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand2</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.join_strands"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.join_strands" title="Link to this definition"></a></dt>
<dd><p>Join two strands together if they match in the right direction and have
a compatible directionality. This also combines the pseudoknot information
and the 3D coordinates of the strands (transforming the coordinates of the
second strand to match the first strand extension). The directionality of
the combined strand is taken from the first strand if it has a sequence,
otherwise it is taken from the second strand. The callbacks of the first
strand are used for the new strand. The strands block of the joined strand
includes the strands blocks of both strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>strand1</strong><span class="classifier">Strand</span></dt><dd><p>The first strand to join.</p>
</dd>
<dt><strong>strand2</strong><span class="classifier">Strand</span></dt><dd><p>The second strand to join.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Optional[Strand]</dt><dd><p>The joined strand if possible, otherwise None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.max_pos">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_pos</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Position</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.max_pos" title="Link to this definition"></a></dt>
<dd><p>The maximum x, y coordinates of the strand in 2D space.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.min_pos">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">min_pos</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Position</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.min_pos" title="Link to this definition"></a></dt>
<dd><p>The minimum x, y coordinates of the strand in 2D space.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.minimal_dimensions">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">minimal_dimensions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Position</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.minimal_dimensions" title="Link to this definition"></a></dt>
<dd><p>The minimum start Positions required to draw the strand in 2D
without reaching negative positions.
Calculating the minimal dimensions doesn’t take into
acount structural errors.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.next_pos">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">next_pos</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Position</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.next_pos" title="Link to this definition"></a></dt>
<dd><p>The next position of the strand: it is the last position plus the
end_direction. Useful for joining two strands.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.pop">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.pop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.pop" title="Link to this definition"></a></dt>
<dd><p>Remove and return a character from the strand at the specified index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>idx</strong><span class="classifier">int</span></dt><dd><p>The index of the character to remove.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The removed character.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.positions">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">positions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Position</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.positions" title="Link to this definition"></a></dt>
<dd><p>A tuple of the positions of each character of the strand in 2D space
(x,y coordinates).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.prec_pos">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">prec_pos</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Position</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.prec_pos" title="Link to this definition"></a></dt>
<dd><p>The preceding position of the strand: it is the start position
minus the direction. Useful for joining two strands.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.reverse">
<span class="sig-name descname"><span class="pre">reverse</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.reverse"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.reverse" title="Link to this definition"></a></dt>
<dd><p>Reverse the directionality of the strand sequence
(5’ to 3’ becomes 3’ to 5’ and vice versa).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Strand</dt><dd><p>The updated strand instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.save_3d_model">
<span class="sig-name descname"><span class="pre">save_3d_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'strand'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.save_3d_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.save_3d_model" title="Link to this definition"></a></dt>
<dd><p>Save the 3D structure of the strand in oxDNA format (.dat, .top)
and optionally as PDB.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename</strong><span class="classifier">str, default=’strand’</span></dt><dd><p>The base name for the output files.</p>
</dd>
<dt><strong>return_text</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, returns the configuration and topology text instead
of saving to file (used for real-time visualization).</p>
</dd>
<dt><strong>pdb</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, exports the structure as a PDB file (requires
<cite>oxDNA_analysis_tools</cite>).</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments passed to the PDB export.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>tuple of str or None</dt><dd><p>Returns (conf_text, top_text) if <cite>return_text</cite> is True, otherwise None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.seq_positions">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seq_positions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Position</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.seq_positions" title="Link to this definition"></a></dt>
<dd><p>The positions of each nucleotide in the strand sequence (x,y coordinates).</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.sequence">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Sequence</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.sequence" title="Link to this definition"></a></dt>
<dd><p>The complete sequence of the strand.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.sequence_list">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_list</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Sequence</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.sequence_list" title="Link to this definition"></a></dt>
<dd><p>The list of consecutive sequences in the strand.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.sequence_slice">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence_slice</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">slice</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.sequence_slice" title="Link to this definition"></a></dt>
<dd><p>The list of slices of the sequence in the strand.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.shift">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift_position</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.shift" title="Link to this definition"></a></dt>
<dd><p>Shift the strand position in the 2D layout.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>shift_position</strong><span class="classifier">Position or Direction or tuple of int</span></dt><dd><p>The shift to apply in x and y direction.</p>
</dd>
<dt><strong>check</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, check if the shift is valid (i.e. not negative coordinates).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Strand</dt><dd><p>The updated strand instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.start">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">start</span></span><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.start" title="Link to this definition"></a></dt>
<dd><p>Start position of the strand in the 2D representation.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.strand">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strand</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.strand" title="Link to this definition"></a></dt>
<dd><p>The strand text characters.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.strands_block">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strands_block</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#pyfurnace.design.core.strand.StrandsBlock" title="pyfurnace.design.core.strand.StrandsBlock"><span class="pre">StrandsBlock</span></a></em><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.strands_block" title="Link to this definition"></a></dt>
<dd><p>The block of strands that has locked 3D coordinates
transformation (their relative position in the 3D space
doesn’t change when joined to other strands).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.Strand.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">transformation_matrix</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#Strand.transform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.Strand.transform" title="Link to this definition"></a></dt>
<dd><p>Apply a transformation matrix to the strand’s 3D coordinates.
Check the Coords class for the transformation matrix format.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>transformation_matrix</strong><span class="classifier">array-like</span></dt><dd><p>The transformation matrix to apply.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.StrandsBlock">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.strand.</span></span><span class="sig-name descname"><span class="pre">StrandsBlock</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#StrandsBlock"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.StrandsBlock" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">set</span></code></p>
<p>A container class that holds a collection of Strand instances.</p>
<p>This class inherits from Python’s built-in list and is designed to manage
a group of Strand objects. It provides methods for adding, removing,
joining, and transforming strands in the block.
All the strands in the block have locked 3D coordinates respect to
each other. This means that when a strand in the block is transformed,
all the other strands in the block are transformed as well with the same
transformation matrix. This is foundamental for managing 3D structures
of motifs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>*args</strong><span class="classifier">Strand</span></dt><dd><p>Strand instances to be added to the block.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">Any</span></dt><dd><p>Additional keyword arguments passed to the list constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>items</strong><span class="classifier">list</span></dt><dd><p>The list of Strand instances stored in the block.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.StrandsBlock.add" title="pyfurnace.design.core.strand.StrandsBlock.add"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add</span></code></a>(item)</p></td>
<td><p>Append a Strand instance to the StrandsBlock.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code></p></td>
<td><p>Remove all elements from this set.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code></p></td>
<td><p>Return a shallow copy of a set.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference</span></code></p></td>
<td><p>Return the difference of two or more sets as a new set.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">difference_update</span></code></p></td>
<td><p>Remove all elements of another set from this set.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">discard</span></code></p></td>
<td><p>Remove an element from a set if it is a member.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection</span></code></p></td>
<td><p>Return the intersection of two sets as a new set.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">intersection_update</span></code></p></td>
<td><p>Update a set with the intersection of itself and another.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">isdisjoint</span></code></p></td>
<td><p>Return True if two sets have a null intersection.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">issubset</span></code>(other, /)</p></td>
<td><p>Test whether every element in the set is in other.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">issuperset</span></code>(other, /)</p></td>
<td><p>Test whether every element in other is in the set.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code></p></td>
<td><p>Remove and return an arbitrary set element.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.StrandsBlock.remove" title="pyfurnace.design.core.strand.StrandsBlock.remove"><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code></a>(item)</p></td>
<td><p>Remove a Strand instance from the StrandsBlock.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference</span></code></p></td>
<td><p>Return the symmetric difference of two sets as a new set.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">symmetric_difference_update</span></code></p></td>
<td><p>Update a set with the symmetric difference of itself and another.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.StrandsBlock.transform" title="pyfurnace.design.core.strand.StrandsBlock.transform"><code class="xref py py-obj docutils literal notranslate"><span class="pre">transform</span></code></a>(T)</p></td>
<td><p>Apply a transformation matrix to all the strands in the StrandsBlock.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">union</span></code></p></td>
<td><p>Return the union of sets as a new set.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.core.strand.StrandsBlock.update" title="pyfurnace.design.core.strand.StrandsBlock.update"><code class="xref py py-obj docutils literal notranslate"><span class="pre">update</span></code></a>(*others[, avoid_strands])</p></td>
<td><p>Update the strands block with the strands in the input iterable.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.StrandsBlock.add">
<span class="sig-name descname"><span class="pre">add</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#StrandsBlock.add"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.StrandsBlock.add" title="Link to this definition"></a></dt>
<dd><p>Append a Strand instance to the StrandsBlock.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>item</strong><span class="classifier">Strand</span></dt><dd><p>The Strand instance to append.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>ValueError</dt><dd><p>If the item is not an instance of Strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.StrandsBlock.remove">
<span class="sig-name descname"><span class="pre">remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#StrandsBlock.remove"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.StrandsBlock.remove" title="Link to this definition"></a></dt>
<dd><p>Remove a Strand instance from the StrandsBlock.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>item</strong><span class="classifier">Strand</span></dt><dd><p>The Strand instance to remove.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>ValueError</dt><dd><p>If the item is not an instance of Strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.StrandsBlock.transform">
<span class="sig-name descname"><span class="pre">transform</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">T</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#StrandsBlock.transform"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.StrandsBlock.transform" title="Link to this definition"></a></dt>
<dd><p>Apply a transformation matrix to all the strands in the StrandsBlock.
For the transformation matrix format, check the Coords class and its
transform method.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>T</strong><span class="classifier">np.ndarray</span></dt><dd><p>The transformation matrix to apply to each strand.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This method assumes that the transformation matrix is compatible
with the strand’s coordinate system.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.core.strand.StrandsBlock.update">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">others</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">avoid_strands</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/strand.html#StrandsBlock.update"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.strand.StrandsBlock.update" title="Link to this definition"></a></dt>
<dd><p>Update the strands block with the strands in the input iterable.
This method is similar to the add method, but it allows for
adding multiple strands at once. It also allows for avoiding
specific strands in the update process.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>*other</strong><span class="classifier">StrandsBlock</span></dt><dd><p>The StrandsBlock instances to update with.</p>
</dd>
<dt><strong>avoid_strands</strong><span class="classifier">set of Strand, default None</span></dt><dd><p>A set of strands to avoid in the update process.</p>
</dd>
<dt><strong>Raises</strong></dt><dd></dd>
<dt><strong>ValueError</strong></dt><dd><p>If any of the input arguments is not a StrandsBlock, or
if avoid_strand is not a Strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="motif">
<h3>Motif<a class="headerlink" href="#motif" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.motif.</span></span><span class="sig-name descname"><span class="pre">Motif</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">strands</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basepair</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">autopairing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lock_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Callback</span></code></p>
<p>Represents a structural motif in an RNA Origami design.</p>
<p>A <cite>Motif</cite> is composed of multiple <cite>Strand</cite> objects, defining an RNA secondary
structure with base pair interactions. It supports coordinate locking,
strand joining, sequence assignment, and 2D motif visualization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>*strands</strong><span class="classifier">list of Strand, optional</span></dt><dd><p>A list of <cite>Strand</cite> objects to initialize the motif.</p>
</dd>
<dt><strong>basepair</strong><span class="classifier">BasePair, dict, optional</span></dt><dd><p>A dictionary defining base pair connections, with <cite>Position</cite> objects as keys
and values representing paired positions.</p>
</dd>
<dt><strong>structure</strong><span class="classifier">str, optional</span></dt><dd><p>A dot-bracket notation representation of the motif’s structure.
If this is provided, the <cite>basepair</cite> parameter is ignored.</p>
</dd>
<dt><strong>autopairing</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, automatically pairs bases based on complementary interactions.
If a basepair or structure is given, autopairing is turned off.</p>
</dd>
<dt><strong>lock_coords</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, locks the coordinate system for all strands in the motif.
When a strand coordinate transformed, all other strands are transformed.</p>
</dd>
<dt><strong>join</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, attempts to merge consecutive strands where possible.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default False</span></dt><dd><p>If True, creates copies of the input strands instead of using references.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the parent <cite>Callback</cite> class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code><span class="classifier">bool</span></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code><span class="classifier">BasePair</span></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code><span class="classifier">dict</span></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code><span class="classifier">bool</span></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code><span class="classifier">tuple of int</span></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code><span class="classifier">tuple of int</span></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code><span class="classifier">int</span></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code><span class="classifier">int</span></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code><span class="classifier">dict</span></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code><span class="classifier">tuple of Position</span></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code><span class="classifier">tuple of Position</span></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code><span class="classifier">Sequence</span></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code><span class="classifier">list of Strand</span></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code><span class="classifier">str</span></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The motif is built from strands and supports strand merging.</p></li>
<li><p>Base pairs can be set manually or inferred automatically.</p></li>
<li><p>Supports transformations, file export, and visual representation.</p></li>
</ul>
<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">align</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">motifs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_junctions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'first'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_shifts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.align"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Align motifs along a given axis by shifting them (without concatenating).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code>], <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>*motifs</strong><span class="classifier">List[Motif]</span></dt><dd><p>List of motifs to align.</p>
</dd>
<dt><strong>axis</strong><span class="classifier">int, default 1</span></dt><dd><p>The numpy axis along which motifs are aligned
(1 for horizontal, 0 for vertical).</p>
</dd>
<dt><strong>extend</strong><span class="classifier">bool, default False</span></dt><dd><p>If True, junctions are extended to accommodate the shift.</p>
</dd>
<dt><strong>align_junctions</strong><span class="classifier">Optional[List[Tuple[int, int]]], default None</span></dt><dd><p>List of tuples specifying which junctions to align.</p>
</dd>
<dt><strong>align_to</strong><span class="classifier">str, default “first”</span></dt><dd><p>Specifies how to align the motifs. Options: “first”, “last”, “center”.
“first” aligns the first junction of the first motif with the first junction
of the second motif. “last” aligns the last junction of the first motif
with the last junction of the second motif. “center” aligns the center
of the first motif with the center of the second motif.</p>
</dd>
<dt><strong>return_shifts</strong><span class="classifier">bool, default False</span></dt><dd><p>If True, returns the shifts applied to each motif during alignment,
along with the aligned motifs (useful for Origami alignment).
If False, only the aligned motifs are returned.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Parse if calling the motifs list explicitly.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>List[Motif]</dt><dd><p>The list of aligned motifs.</p>
</dd>
<dt>List[Tuple[int, int]], optional</dt><dd><p>The list of shifts applied to each motif during alignment.
Only returned if return_shifts is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.append"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Append a strand to the motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>strand</strong><span class="classifier">Strand</span></dt><dd><p>The strand to be added.</p>
</dd>
<dt><strong>join</strong><span class="classifier">bool, default True</span></dt><dd><p>Whether to attempt joining the new strand with existing strands.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default False</span></dt><dd><p>If True, a copy of the strand is appended.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The updated motif.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the provided <cite>strand</cite> is not a <cite>Strand</cite> instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">autopairing</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em></dt>
<dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">basepair</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">BasePair</span></em></dt>
<dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">concat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">motifs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_based</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_junctions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">unlock_strands</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_shifts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.concat"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Concatenate multiple motifs along a specified axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>*motifs</strong><span class="classifier">list of Motif</span></dt><dd><p>List of motifs to concatenate.</p>
</dd>
<dt><strong>axis</strong><span class="classifier">int, default 1</span></dt><dd><p>The numpy axis along which motifs are aligned.
1 for horizontal alignment, 0 for vertical alignment.</p>
</dd>
<dt><strong>extend</strong><span class="classifier">bool, default False</span></dt><dd><p>Whether to extend junctions before concatenation.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default True</span></dt><dd><p>Whether to copy motifs before concatenation.</p>
</dd>
<dt><strong>align</strong><span class="classifier">bool, default True</span></dt><dd><p>Whether to align motifs before concatenation.</p>
</dd>
<dt><strong>position_based</strong><span class="classifier">bool, default False</span></dt><dd><p>Whether to align motifs based on positions instead of junctions.</p>
</dd>
<dt><strong>align_junctions</strong><span class="classifier">list of tuple, optional</span></dt><dd><p>List of junction indices for alignment.</p>
</dd>
<dt><strong>unlock_strands</strong><span class="classifier">bool, default False</span></dt><dd><p>Whether to unlock all strands, so they are not part
of the same Strand block (different coordinate systems).</p>
</dd>
<dt><strong>return_shifts</strong><span class="classifier">bool, default False</span></dt><dd><p>Whether to return the shift values for each motif to be aligned
and concatenated. If True, a list of tuples with the shifts is returned,
along with the concatenated motif (useful for Origami alignment).</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the Motif constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The concatenated Motif object.</p>
</dd>
<dt>List[Tuple[int, int]], optional</dt><dd><p>The list of shifts applied to each motif during alignment.
Only returned if return_shifts is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">callback</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Return a copy of the motif.</p>
<p>This method creates a custom deep copy of the current motif, preserving
its strand arrangements, base pairings, and additional attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>callback</strong><span class="classifier">callable, default None</span></dt><dd><p>A callback function to be registered in the copied motif.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>If the keyword ‘deepcopy’ with a list of attributes names is passed,
the attributes will deepcopyed instead referenced.
Same for the keyword ‘copy’ with a list of attributes names.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif or subclass</dt><dd><p>A new instance of the motif or its subclass, with the same properties
and attributes as the original.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">copy_strands_preserve_blocks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strands</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.copy_strands_preserve_blocks"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Copy a list of strands while preserving strand blocks.
Ensures that strands that belong to the same motif remain linked together.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>strands</strong><span class="classifier">list of Strand</span></dt><dd><p>The list of strands to be copied.</p>
</dd>
<dt><strong>motif</strong><span class="classifier">Motif, optional</span></dt><dd><p>If provided, associates the strands with the given motif and registers
callback functions.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>list of Strand</dt><dd><p>A list of copied strands.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">extend_junctions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">skip_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">skip_directions</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">until</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.extend_junctions"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Extend the junctions of the motif in the direction of the junctions.
Unless the skip_axis is specified, the junctions are extended
in all directions. The until parameter allows to specify
the position until which to extend the junctions in the positive direction.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>skip_axis: int, optional</strong></dt><dd><p>The numpy axis to skip when extending the junctions
(1 for horizontal, 0 for vertical).</p>
</dd>
<dt><strong>skip_directions: List[Direction], optional</strong></dt><dd><p>The list of directions to skip when extending the junctions.</p>
</dd>
<dt><strong>until</strong><span class="classifier">Position, optional</span></dt><dd><p>The position until which to extend the junctions
If None, extend until the end of the motif in all directions
(to the origin and to the maximum position).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The motif with the extended junctions.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the axis parameter is not None, 0 or 1.</p>
</dd>
<dt>MotifStructureError</dt><dd><p>If there is an error in extending the junction</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">flip</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">horizontally</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vertically</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reorder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.flip"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Flip the strands of the motif horizontally and/or vertically inplace.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>horizontally</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, flip the strands horizontally.</p>
</dd>
<dt><strong>vertically</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, flip the strands vertically.</p>
</dd>
<dt><strong>strand_index</strong><span class="classifier">list, default None</span></dt><dd><p>The list of indices of the strands to flip.
If None, all the strands are flipped.</p>
</dd>
<dt><strong>reorder</strong><span class="classifier">bool, default False</span></dt><dd><p>If True, reorder the strands after flipping,
so that the first strand is the one with the 5’ end,
then the strand with the lowest y start position,
and finally the strand with the lowest x start position.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The motif with the flipped strands.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">folding_barriers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kl_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em><span class="sig-paren">)</span></dt>
<dd><p>Compute the folding barriers for a given RNA secondary structure.
This function is based on
ROAD: <a class="reference external" href="https://www.nature.com/articles/s41557-021-00679-1">https://www.nature.com/articles/s41557-021-00679-1</a>
This function analyzes the dot-bracket representation of the RNA secondary
structure to determine folding barriers based on base pair topology and
kinetic delay.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">str</span></dt><dd><p>The dot-bracket notation representing the secondary structure of the RNA.
If folding barriers is called from a Motif or Origami, the structure is
already provided by the object in the dot-bracket format.</p>
</dd>
<dt><strong>kl_delay</strong><span class="classifier">int, optional, default=150</span></dt><dd><p>The number of nucleotides (nts) of delay before kissing loops (KLs) snap
closed. A typical realistic value is around 350 (~1 second), while a
more conservative  setting is 150 by default.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Tuple[str, int]</dt><dd><p>A tuple containing:
- A string representing the barrier map where:
‘─’ = no barrier (0 penalty),
‘▂’ = opening pair weak barrier (1 penalty),
‘▄’ = cloding pair weak barrier (1 penalty),
‘█’ = strong barrier (2 penalty).
- An integer score indicating the total penalty based on barrier strengths.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function assigns barrier strengths based on the topology of base pairs
and kinetic constraints, ensuring proper folding predictions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.from_file"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Create a Motif object from a text file containing a motif sketch.
Each Strand is read starting from the <cite>5</cite> symbol. If you want to add the 5’
terminal symbol, start the strand with <cite>55</cite>. Only one symbol should be placed
next to the <cite>5</cite> start of the strand in order to guess the right direction.
Alternatively, you can use symbols <cite>^v&gt;&lt;</cite> to start a strand and indicate
the start direction (up, down, right, left); in this case you can place
multiple symbols next to the start symbol.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>file_path</strong><span class="classifier">str</span></dt><dd><p>Path to the file containing the motif structure.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the Motif constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The constructed Motif object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_list</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.from_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Create a Motif object from a list of strings representing a motif sketch.
Each Strand is read starting from the <cite>5</cite> symbol. If you want to add the 5’
terminal symbol, start the strand with <cite>55</cite>. Only one symbol should be placed
next to the <cite>5</cite> start of the strand in order to guess the right direction.
Alternatively, you can use symbols <cite>^v&gt;&lt;</cite> to start a strand and indicate
the start direction (up, down, right, left); in this case you can place
multiple symbols next to the start symbol.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>motif_list</strong><span class="classifier">list of str</span></dt><dd><p>The list of strings representing the motif.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the Motif constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The constructed Motif object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_denergy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.from_structure"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Parse a structure or sequence representation to a Motif object.
If a structure is not provided, it is calculated from the sequence
with RNAfold. If a sequence is not provided, it is assumed to be
a sequence of ‘N’s of the same length as the structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">Union[str, dict, BasePair, Node]</span></dt><dd><p>The structure representation to convert.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>The sequence or sequence constraints of the motif.</p>
</dd>
<dt><strong>pk_energy</strong><span class="classifier">float, optional</span></dt><dd><p>The energy of the pseudoknots (if present).</p>
</dd>
<dt><strong>pk_denergy</strong><span class="classifier">float, optional</span></dt><dd><p>The energy tolerance of the pseudoknots (if present).</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the Motif constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The Motif object created from the structure representation.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_text</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.from_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Create a Motif object from a text string representing a motif sketch.
Each Strand is read starting from the <cite>5</cite> symbol. If you want to add the 5’
terminal symbol, start the strand with <cite>55</cite>. Only one symbol should be placed
next to the <cite>5</cite> start of the strand in order to guess the right direction.
Alternatively, you can use symbols <cite>^v&gt;&lt;</cite> to start a strand and indicate
the start direction (up, down, right, left); in this case you can place
multiple symbols next to the start symbol.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>motif_text</strong><span class="classifier">str</span></dt><dd><p>The motif structure in text format.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the Motif constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The constructed Motif object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_position_map</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.get_position_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Get a dictionary of positions as keys and the corresponding
characters as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Dict[Tuple[int, int], str]</dt><dd><p>A dictionary mapping positions to characters.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_sequential_shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motifs</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">position_based</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.get_sequential_shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Calculate the shift needed to align motifs sequentially.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>motifs</strong><span class="classifier">List[Motif]</span></dt><dd><p>List of motifs to align.</p>
</dd>
<dt><strong>*args: List[Motif]</strong></dt><dd><p>Additional motifs to align.</p>
</dd>
<dt><strong>axis</strong><span class="classifier">int, default 1</span></dt><dd><p>The numpy axis along which the shift is computed
(1 for horizontal, 0 for vertical).</p>
</dd>
<dt><strong>position_based</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, the shift is calculated based on motif positions
rather than junctions.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>List[int]</dt><dd><p>A list of shift integers for each motif along the axis.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>m1:
    --NN--
      ::
    --NN--
m2:
    --SK--
      ::
    --SK--
Resulting shift: 2
    [0, 2]
To have:
    --NN----SK--
      ::    ::
    --NN----SK--
</pre></div>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_strand_index_map</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.get_strand_index_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Get a dictionary of positions as keys and the index of the
strand in the motif as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Dict</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Dict[Tuple[int, int], int]</dt><dd><p>A dictionary mapping positions to strand indexes</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strand</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.insert"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Insert a strand at a given index in the motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>index</strong><span class="classifier">int</span></dt><dd><p>The index at which to insert the strand.</p>
</dd>
<dt><strong>strand</strong><span class="classifier">Strand</span></dt><dd><p>The strand to be inserted.</p>
</dd>
<dt><strong>join</strong><span class="classifier">bool, default True</span></dt><dd><p>Whether to attempt joining the new strand with existing strands.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default False</span></dt><dd><p>If True, a copy of the strand is inserted.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The updated motif.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If <cite>strand</cite> is not a <cite>Strand</cite> instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">join_strands</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strands</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.join_strands"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Attempt to join consecutive strands and return the list of joined strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>strands</strong><span class="classifier">List[Strand]</span></dt><dd><p>The list of strands to be joined.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>List[Strand]</dt><dd><p>The list of joined strands.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">junctions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1),
left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lock_coords</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em></dt>
<dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">max_pos</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">min_pos</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_char</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em></dt>
<dd><p>The maximum length of the motif lines.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_lines</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em></dt>
<dd><p>The number of lines in the motif structure.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pair_map</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="w"> </span><span class="p"><span class="pre">|</span></span><span class="w"> </span><span class="pre">None</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The dictionary of the paired indexes (alternative to the dot bracket
notation).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.pop"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Remove and return the strand at the specified index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><code class="xref py py-class docutils literal notranslate"><span class="pre">Strand</span></code></a></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>index</strong><span class="classifier">int, default -1</span></dt><dd><p>The index of the strand to remove.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Strand</dt><dd><p>The removed strand.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">positions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Position</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>A tuple of the positions of each character of each strand in 2D space
(x,y coordinates).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">replace_all_strands</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strands</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.replace_all_strands"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Replace all the strands in the motif with the provided list of strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>strands</strong><span class="classifier">List[Strand]</span></dt><dd><p>The list of strands to replace the current strands.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default True</span></dt><dd><p>If True, the strands are copied before replacing the current strands.</p>
</dd>
<dt><strong>join</strong><span class="classifier">bool, default False</span></dt><dd><p>Whether to attempt joining the new strands.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The updated motif.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If any of the provided strands is not a <cite>Strand</cite> instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">rotate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">times</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.rotate"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Rotate the motif 90 degrees clockwise.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>times</strong><span class="classifier">int, default 1</span></dt><dd><p>The number of times to rotate the motif.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The rotated motif.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_3d_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'motif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">topology</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forces</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_forces</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.save_3d_model"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Save the motif 3D structure in a file for structural 3D modeling.
This function save the 3D oxDNA motif representation in conformation and
topology xoDNA-format files, and optionally in PDB format.
It can also save additional force constraints for oxDNA simulation
and protein configurations if specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename</strong><span class="classifier">str, optional, default=”motif”</span></dt><dd><p>The base filepath for the generated oxDNA files (without extension).</p>
</dd>
<dt><strong>config</strong><span class="classifier">bool, default=True</span></dt><dd><p>If True, generates a configuration (.dat) file with nucleotide positions.</p>
</dd>
<dt><strong>topology</strong><span class="classifier">bool, default=True</span></dt><dd><p>If True, generates a topology (.top) file specifying strand connectivity.</p>
</dd>
<dt><strong>forces</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, saves force constraints for base-pair interactions.</p>
</dd>
<dt><strong>pk_forces</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, saves force constraints specifically for pseudoknots.</p>
</dd>
<dt><strong>return_text</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, returns the generated oxDNA configuration and topology as strings
instead of writing to files (used for real-time visualization).</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>If provided, uses the given sequence to generate the topology.</p>
</dd>
<dt><strong>pdb</strong><span class="classifier">bool, optional, default=False</span></dt><dd><p>If True, exports a PDB (Protein Data Bank) file for visualization.</p>
</dd>
<dt><strong>**kwargs</strong></dt><dd><p>Additional arguments for customizing force constraints
and PDB export settings.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Optional[Tuple[str, str]]</dt><dd><p>If <cite>return_text</cite> is True, returns a tuple containing:
- The configuration file content as a string.
- The topology file content as a string.
Otherwise, returns None.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function requires the oxDNA analysis tools package to be installed
for the generation of force constraints and pdb export.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'motif'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.save_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Save the motif sequences in a FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename</strong><span class="classifier">str, optional</span></dt><dd><p>The filepath to save (without extension). Default is ‘motif’.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'motif'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.save_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Save the motif representation as a text file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename_path</strong><span class="classifier">str, optional</span></dt><dd><p>The filepath to save (without extension). Default is ‘motif’.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seq_positions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Position</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).
The sequence has always the directionality 5’ to 3’</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em></dt>
<dd><p>Return the sequence of the motif.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift_vect</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extend</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Shift the motif of the given shift vector.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>shift_vect</strong><span class="classifier">Tuple[int, int]</span></dt><dd><p>The (x, y) shift values.</p>
</dd>
<dt><strong>extend</strong><span class="classifier">bool, default False</span></dt><dd><p>Whether to extend junctions when shifting
(in the direction opposite to the shifting direction).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The shifted motif.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If shifting moves strands to negative positions.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">sort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.sort"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Sort the strands in the motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>key</strong><span class="classifier">function, optional</span></dt><dd><p>The function to use to sort the strands. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, the strands are
sorted by:</p>
<ul class="simple">
<li><p>strands with a 5’ end first</p></li>
<li><p>lowest starting y position</p></li>
<li><p>lowest starting x position</p></li>
</ul>
</dd>
<dt><strong>reverse</strong><span class="classifier">bool, default False</span></dt><dd><p>If <code class="docutils literal notranslate"><span class="pre">True</span></code>, sort in descending order.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The motif with the sorted strands.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strands</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><span class="pre">Strand</span></a><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The list of strands in the motif.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">strip</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">skip_axis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.strip"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Remove the empty lines/columns in the motif structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>skip_axis: int, optional</strong></dt><dd><p>The numpy axis to skip when stripping the motif
(1 for horizontal, 0 for vertical).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The stripped motif.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">structure</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em></dt>
<dd><p>Return the dot bracket representation of the motif.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">static</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">trace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pos</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limits</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/motif.html#Motif.trace"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Trace strands in a motif representation in a recursive manner.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>motif_list</strong><span class="classifier">list of str</span></dt><dd><p>The list of strings of the motif sketch.</p>
</dd>
<dt><strong>pos</strong><span class="classifier">tuple of int</span></dt><dd><p>The starting position.</p>
</dd>
<dt><strong>direction</strong><span class="classifier">tuple of int</span></dt><dd><p>The tracing direction.</p>
</dd>
<dt><strong>limits</strong><span class="classifier">tuple of int</span></dt><dd><p>The boundary limits.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The traced strand sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="origami">
<h3>Origami<a class="headerlink" href="#origami" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.origami.</span></span><span class="sig-name descname"><span class="pre">Origami</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matrix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'left'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ss_assembly</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Callback</span></code></p>
<p>A class for building and manipulating RNA Origami structures.</p>
<p>The Origami class organizes RNA <cite>Motif</cite> objects into a 2D matrix to
represent spatial arrangements of strands. It supports stacking motifs
horizontally and vertically, calculating vertical junctions and connections,
assembling the full origami structure in a motif object, and exporting
the full structure and sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>matrix</strong><span class="classifier">Motif or list of Motif or list of list of Motif</span></dt><dd><p>A motif or 1D/2D list of Motifs to initialize the origami layout.</p>
</dd>
<dt><strong>*args</strong><span class="classifier">Motif or list of Motif</span></dt><dd><p>Additional motifs or rows of motifs to add to the matrix.</p>
</dd>
<dt><strong>align</strong><span class="classifier">{‘left’, ‘first’, ‘center’}, default=’left’</span></dt><dd><p>How motifs should be aligned vertically.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default=False</span></dt><dd><p>Whether to create a copy of the motifs before adding them to the origami.</p>
</dd>
<dt><strong>ss_assembly</strong><span class="classifier">bool, default=False</span></dt><dd><p>Wether to assemble the 3D structure of the origami without locking the
coordinates of the motifs.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments passed to the Callback base class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code><span class="classifier">str</span></dt><dd><p>The alignment type of the rows of the origami.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">assembled</span></code><span class="classifier">Motif</span></dt><dd><p>The matrix of the origami with the motif shifted in the correct position for the assembly.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code><span class="classifier">List[int]</span></dt><dd><p>The number of characters in each line of the origami.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code><span class="classifier">int</span></dt><dd><p>The number of lines in the origami.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_motifs</span></code><span class="classifier">int</span></dt><dd><p>The number of motifs in the origami.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code><span class="classifier">dict</span></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pos_index_map</span></code><span class="classifier">Dict[Position, Tuple[int, int]]</span></dt><dd><p>A dictionary with the symbols position (x, y) as keys and the matrix index (y, x) of the motif that contains it as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">index_shift_map</span></code><span class="classifier">Dict[Position, Tuple[int, int]]</span></dt><dd><p>A dictionary with the slice of the motif in the matrix as key (y, x) and positional shift of the motif as values (y, x).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pseudoknots</span></code><span class="classifier">dict</span></dt><dd><p>A dictionary with the pseudoknot information.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code><span class="classifier">Sequence</span></dt><dd><p>The sequence of the origami, as a Sequence.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code><span class="classifier">Tuple[Position]</span></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">ss_assembly</span></code><span class="classifier">bool</span></dt><dd><p>Boolean indicating if the origami 3d structure is assembled without locking the coordinates of the motifs.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code><span class="classifier">List[Strand]</span></dt><dd><p>The strands of the origami.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code><span class="classifier">str</span></dt><dd><p>The dot-bracket structure of the origami.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(item[, copy])</p></td>
<td><p>Append a Motif or a list of Motifs to the end of the matrix.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">barrier_repr</span></code>([kl_delay, barriers, return_list])</p></td>
<td><p>Overlay folding barrier characters onto the structure visualization.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>()</p></td>
<td><p>Create a deep copy of the Origami.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">duplicate_line</span></code>(idx[, insert_idx])</p></td>
<td><p>Duplicate a line of motifs and optionally insert it elsewhere.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to an Origami object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_motif_at_position</span></code>(position)</p></td>
<td><p>Get a motif from its position in 2D coordinates.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_motif_at_seq_index</span></code>(index)</p></td>
<td><p>Get the motif that contains the given sequence index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_motif_type</span></code>(motif_type)</p></td>
<td><p>Get all motifs in the Origami that match the given type.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_slice_at_seq_index</span></code>(index)</p></td>
<td><p>Get matrix coordinates of the motif containing the given sequence index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">improve_folding_pathway</span></code>([kl_delay])</p></td>
<td><p>Suggest a better folding pathway by circularly shifting the structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">index</span></code>(condition)</p></td>
<td><p>Find the matrix coordinates of motifs that satisfy a given condition.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(idx, item[, copy])</p></td>
<td><p>Insert a Motif or list of Motifs at a specific position.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>(idx)</p></td>
<td><p>Remove and return a motif or line of motifs at the given index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reload</span></code>()</p></td>
<td><p>Recompute the internal structure and regenerate the assembled motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">remove</span></code>(motif)</p></td>
<td><p>Remove a specific motif from the matrix.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>(filename_path[, return_text])</p></td>
<td><p>Save the sequence of the Origami to a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>(filename_path[, to_road, return_text])</p></td>
<td><p>Save only the structure part of the Origami to a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">to_road</span></code>()</p></td>
<td><p>Try to convert the Origami's text representation into ROAD-compatible format.</p></td>
</tr>
</tbody>
</table>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">Motif</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">Strand</span></code>, <code class="xref py py-obj docutils literal notranslate"><span class="pre">Sequence</span></code></dt><dd></dd>
</dl>
</div>
<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">align</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Literal</span><span class="p"><span class="pre">[</span></span><span class="s"><span class="pre">'left'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'first'</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="s"><span class="pre">'center'</span></span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The alignment type of the rows of the origami.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">append</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">item</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.append"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Append a Motif or a list of Motifs to the end of the matrix.
If the item is a single Motif, it is appended to the last line of the matrix.
If the item is a list of Motifs, it is appended as a new line in the matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>item</strong><span class="classifier">Motif or list of Motif</span></dt><dd><p>The motif(s) to append.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default=True</span></dt><dd><p>Whether to copy motifs before appending.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>TypeError</dt><dd><p>If <cite>item</cite> is not a Motif or a list of Motifs.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">assembled</span></span></dt>
<dd><p>The matrix of the origami with the motif shifted in the correct position
for the assembly. The assembled matrix contains rows with the vertical
connection motifs.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">barrier_repr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kl_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barriers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.barrier_repr"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Overlay folding barrier characters onto the structure visualization.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>kl_delay</strong><span class="classifier">int, default=150</span></dt><dd><p>Delay parameter for computing folding barriers.</p>
</dd>
<dt><strong>barriers</strong><span class="classifier">str, optional</span></dt><dd><p>Precomputed folding barrier string. If None, it will be recomputed.</p>
</dd>
<dt><strong>return_list</strong><span class="classifier">bool, default=False</span></dt><dd><p>Whether to return the result as a list of lines instead of a string.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str or list of str</dt><dd><p>The annotated structure as a single string or a list of lines.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Create a deep copy of the Origami.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Origami</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Origami</dt><dd><p>A new instance identical to the current one.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">duplicate_line</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">insert_idx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.duplicate_line"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Duplicate a line of motifs and optionally insert it elsewhere.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>idx</strong><span class="classifier">int</span></dt><dd><p>Index of the line to duplicate.</p>
</dd>
<dt><strong>insert_idx</strong><span class="classifier">int, optional</span></dt><dd><p>Line index at which to insert the duplicated line. If None, it will be
added at the end.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>ValueError</dt><dd><p>If the given <cite>idx</cite> is not an integer.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">folding_barriers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kl_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em><span class="sig-paren">)</span></dt>
<dd><p>Compute the folding barriers for a given RNA secondary structure.
This function is based on
ROAD: <a class="reference external" href="https://www.nature.com/articles/s41557-021-00679-1">https://www.nature.com/articles/s41557-021-00679-1</a>
This function analyzes the dot-bracket representation of the RNA secondary
structure to determine folding barriers based on base pair topology and
kinetic delay.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">str</span></dt><dd><p>The dot-bracket notation representing the secondary structure of the RNA.
If folding barriers is called from a Motif or Origami, the structure is
already provided by the object in the dot-bracket format.</p>
</dd>
<dt><strong>kl_delay</strong><span class="classifier">int, optional, default=150</span></dt><dd><p>The number of nucleotides (nts) of delay before kissing loops (KLs) snap
closed. A typical realistic value is around 350 (~1 second), while a
more conservative  setting is 150 by default.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Tuple[str, int]</dt><dd><p>A tuple containing:
- A string representing the barrier map where:
‘─’ = no barrier (0 penalty),
‘▂’ = opening pair weak barrier (1 penalty),
‘▄’ = cloding pair weak barrier (1 penalty),
‘█’ = strong barrier (2 penalty).
- An integer score indicating the total penalty based on barrier strengths.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function assigns barrier strengths based on the topology of base pairs
and kinetic constraints, ensuring proper folding predictions.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">classmethod</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_structure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_denergy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.from_structure"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Parse a structure or sequence representation to an Origami object.
If a structure is not provided, it is calculated from the sequence
with RNAfold. If a sequence is not provided, it is assumed to be
a sequence of ‘N’s of the same length as the structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Origami</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">Union[str, dict, BasePair, Node]</span></dt><dd><p>The structure representation to convert.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>The sequence or sequence constraints of the motif.</p>
</dd>
<dt><strong>pk_energy</strong><span class="classifier">float, optional</span></dt><dd><p>The energy of the pseudoknots (if present).</p>
</dd>
<dt><strong>pk_denergy</strong><span class="classifier">float, optional</span></dt><dd><p>The energy tolerance of the pseudoknots (if present).</p>
</dd>
<dt><strong>motif_list</strong><span class="classifier">List[Motif], optional</span></dt><dd><p>A list of specific motifs to parse the structure.
By default, the motifs are stems and aptamers.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments to pass to the Motif constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>The Origami object created from the structure representation.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_motif_at_position</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">position</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.get_motif_at_position"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Get a motif from its position in 2D coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>position</strong><span class="classifier">tuple of int</span></dt><dd><p>Global coordinate (x, y) in the assembled structure.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>The motif located at the position.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the position is not valid or not found in the map.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_motif_at_seq_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.get_motif_at_seq_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Get the motif that contains the given sequence index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>index</strong><span class="classifier">int</span></dt><dd><p>Sequence index in the full assembled sequence.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif</dt><dd><p>Motif containing the base at the given index.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_motif_type</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.get_motif_type"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Get all motifs in the Origami that match the given type.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>motif_type</strong><span class="classifier">type</span></dt><dd><p>Motif subclass/type to filter.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>List[Motif]</dt><dd><p>All motifs of the specified type.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_slice_at_seq_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.get_slice_at_seq_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Get matrix coordinates of the motif containing the given sequence index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>index</strong><span class="classifier">int</span></dt><dd><p>Index in the full sequence.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Tuple[int, int]</dt><dd><p>(row, column) coordinates in the matrix.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the index is not found.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">improve_folding_pathway</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">kl_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.improve_folding_pathway"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Suggest a better folding pathway by circularly shifting the structure.
This method attempts to find a better folding pathway by shifting the
structure to a new position. IMPORTANT: this method is designed for
simple origami blueprints based on DAE crossovers and may not work
correctly for different structures.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Origami</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>kl_delay</strong><span class="classifier">int, default=150</span></dt><dd><p>Delay parameter for kinetic loop folding.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>A new Origami object with an optimized folding pathway.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">condition</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.index"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Find the matrix coordinates of motifs that satisfy a given condition.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>condition</strong><span class="classifier">Callable[[Motif], bool] or Motif</span></dt><dd><p>A function that takes a Motif and returns True if it matches,
or a Motif instance to match directly.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>List[Tuple[int, int]]</dt><dd><p>List of (row, column) indices of matching motifs.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If <cite>condition</cite> is neither a callable nor a Motif.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">index_shift_map</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em></dt>
<dd><p>A dictionary with the slice of the motif in the matrix as key (y, x)
and positional shift of the motif as values (y, x). The shift is the
difference between the position of the motif in the matrix and the position
of the motif in the assembled origami.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">insert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">item</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.insert"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Insert a Motif or list of Motifs at a specific position.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>idx</strong><span class="classifier">int, slice, tuple of int</span></dt><dd><p>The index or coordinate at which to insert the item(s).</p>
</dd>
<dt><strong>item</strong><span class="classifier">Motif or list of Motif</span></dt><dd><p>The motif(s) to insert.</p>
</dd>
<dt><strong>copy</strong><span class="classifier">bool, default=True</span></dt><dd><p>Whether to copy motifs before inserting.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>ValueError</dt><dd><p>If the index or item is invalid.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_char</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The number of characters in each line of the origami.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_lines</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em></dt>
<dd><p>The number of lines in the origami.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">num_motifs</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">int</span></em></dt>
<dd><p>The number of motifs in the origami.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pair_map</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em></dt>
<dd><p>The dictionary of the paired indexes (alternative to the dot bracket
notation).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.pop"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Remove and return a motif or line of motifs at the given index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Motif</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>idx</strong><span class="classifier">int, slice, tuple of int</span></dt><dd><p>The index or coordinates to remove from.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Motif or list of Motif</dt><dd><p>The removed motif(s).</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the index is not valid.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pos_index_map</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em></dt>
<dd><p>A dictionary with the symbols position (x, y) as keys and the
matrix index (y, x) of the motif that contains it as values.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">positions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">Position</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The positions of the characters in the origami. Is the same as
calling the assembled.motif.positions.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pseudoknots</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em></dt>
<dd><p>A dictionary with the pseudoknot information.
The dictionary has pseudoknot IDs as keys and the pseudoknot information as
values.
The pseudoknot information is a dictionary with the following keys:</p>
<blockquote>
<div><ul class="simple">
<li><dl class="simple">
<dt>ind_fwd: a list of tuples (start, end) with the indices of the forward</dt><dd><p>sequences of the pseudoknot</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>ind_rev: a list of tuples (start, end) with the indices of the reverse</dt><dd><p>sequences of the pseudoknot</p>
</dd>
</dl>
</li>
<li><p>E: the energy of the pseudoknot</p></li>
<li><p>dE: the energy tolerance of the pseudoknot</p></li>
</ul>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">reload</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.reload"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Recompute the internal structure and regenerate the assembled motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">remove</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.remove"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Remove a specific motif from the matrix.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>motif</strong><span class="classifier">Motif</span></dt><dd><p>The motif to remove.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_3d_model</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'motif'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">config</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">topology</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forces</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_forces</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pdb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span></dt>
<dd><p>Save the motif 3D structure in a file for structural 3D modeling.
This function save the 3D oxDNA motif representation in conformation and
topology xoDNA-format files, and optionally in PDB format.
It can also save additional force constraints for oxDNA simulation
and protein configurations if specified.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename</strong><span class="classifier">str, optional, default=”motif”</span></dt><dd><p>The base filepath for the generated oxDNA files (without extension).</p>
</dd>
<dt><strong>config</strong><span class="classifier">bool, default=True</span></dt><dd><p>If True, generates a configuration (.dat) file with nucleotide positions.</p>
</dd>
<dt><strong>topology</strong><span class="classifier">bool, default=True</span></dt><dd><p>If True, generates a topology (.top) file specifying strand connectivity.</p>
</dd>
<dt><strong>forces</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, saves force constraints for base-pair interactions.</p>
</dd>
<dt><strong>pk_forces</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, saves force constraints specifically for pseudoknots.</p>
</dd>
<dt><strong>return_text</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, returns the generated oxDNA configuration and topology as strings
instead of writing to files (used for real-time visualization).</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>If provided, uses the given sequence to generate the topology.</p>
</dd>
<dt><strong>pdb</strong><span class="classifier">bool, optional, default=False</span></dt><dd><p>If True, exports a PDB (Protein Data Bank) file for visualization.</p>
</dd>
<dt><strong>**kwargs</strong></dt><dd><p>Additional arguments for customizing force constraints
and PDB export settings.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Optional[Tuple[str, str]]</dt><dd><p>If <cite>return_text</cite> is True, returns a tuple containing:
- The configuration file content as a string.
- The topology file content as a string.
Otherwise, returns None.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>This function requires the oxDNA analysis tools package to be installed
for the generation of force constraints and pdb export.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.save_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Save the sequence of the Origami to a FASTA file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename_path</strong><span class="classifier">str</span></dt><dd><p>Path to the output file.</p>
</dd>
<dt><strong>return_text</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, return the text instead of saving it to a file.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Optional[str]</dt><dd><p>The FASTA text if return_text is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">save_text</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_road</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_text</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.save_text"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Save only the structure part of the Origami to a text file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>filename_path</strong><span class="classifier">str</span></dt><dd><p>Path to the output file.</p>
</dd>
<dt><strong>to_road</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, convert to ROAD-compatible format.</p>
</dd>
<dt><strong>return_text</strong><span class="classifier">bool, default=False</span></dt><dd><p>If True, return the text instead of saving it to a file.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Optional[str]</dt><dd><p>The text if return_text is True.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">seq_positions</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">Position</span><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).
The sequence has always the directionality 5’ to 3’</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">sequence</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">Sequence</span></em></dt>
<dd><p>The sequence of the origami, as a Sequence.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ss_assembly</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">bool</span></em></dt>
<dd><p>Boolean indicating if the origami 3d structure
is assembled without locking the coordinates of the motifs.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">strands</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><a class="reference internal" href="#pyfurnace.design.core.strand.Strand" title="pyfurnace.design.core.strand.Strand"><span class="pre">Strand</span></a><span class="p"><span class="pre">]</span></span></em></dt>
<dd><p>The strands of the origami.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">structure</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em></dt>
<dd><p>The dot-bracket structure of the origami.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">to_road</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/origami.html#Origami.to_road"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Try to convert the Origami’s text representation into ROAD-compatible format.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>str</dt><dd><p>ROAD-compatible structure representation.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="sequence">
<h3>Sequence<a class="headerlink" href="#sequence" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.sequence.</span></span><span class="sig-name descname"><span class="pre">Sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directionality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'53'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">Callback</span></code></p>
<p>Represents a nucleotide sequence with 5’ to 3’ or 3’ to 5’ directionality.</p>
<p>This class provides functionality for manipulating RNA sequences
with support for slicing, translation, complementarity, GC content,
and integration with structure-aware tools.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>The nucleotide sequence.</p>
</dd>
<dt><strong>directionality</strong><span class="classifier">{‘53’, ‘35’}, optional</span></dt><dd><p>Directionality of the sequence (default is ‘53’).</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments passed to the Callback base class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">directionality</span></code><span class="classifier">str</span></dt><dd><p>directionality of the sequence (either ‘53’ or ‘35’)</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complement</span></code>()</p></td>
<td><p>Return the complement of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>(**kwargs)</p></td>
<td><p>Create a copy of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">distance</span></code>(other)</p></td>
<td><p>Compute the number of mismatched bases between this sequence and another.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">find</span></code>(sub)</p></td>
<td><p>Find the first occurrence of a subsequence in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_repeated_subsequence</span></code>([min_length])</p></td>
<td><p>Find all subsequences of minimum length that appear more than once.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">gc_content</span></code>([extended_alphabet])</p></td>
<td><p>Calculate the GC content of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_random_sequence</span></code>([structure])</p></td>
<td><p>Generate a randomized sequence compatible with the IUPAC symbols and optionally a dot-bracket structure to be respected.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">lower</span></code>()</p></td>
<td><p>Convert sequence to lowercase.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">molecular_weight</span></code>()</p></td>
<td><p>Calculate the molecular weight of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>(idx)</p></td>
<td><p>Remove and return a nucleotide at the given index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace</span></code>(old, new)</p></td>
<td><p>Replace all occurrences of <cite>old</cite> with <cite>new</cite> in the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse</span></code>([inplace])</p></td>
<td><p>Reverse the directionality of the sequence.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">reverse_complement</span></code>()</p></td>
<td><p>Return the reverse complement of the sequence.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">split</span></code>([sep])</p></td>
<td><p>Split the sequence by a separator.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">translate</span></code>(dictionary[, inplace])</p></td>
<td><p>Translate the sequence using a mapping dictionary.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">upper</span></code>()</p></td>
<td><p>Convert sequence to uppercase.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Return the complement of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Sequence</dt><dd><p>Complementary sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Create a copy of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Sequence</dt><dd><p>A new Sequence instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">directionality</span></span></dt>
<dd><p>directionality of the sequence (either ‘53’ or ‘35’)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">other</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Compute the number of mismatched bases between this sequence and another.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>other</strong><span class="classifier">str or Sequence</span></dt><dd><p>Sequence to compare against.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd><p>Number of incompatible positions.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the input is invalid or lengths do not match.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">find</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sub</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.find"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Find the first occurrence of a subsequence in the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>sub</strong><span class="classifier">str</span></dt><dd><p>Subsequence to find.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd><p>Index of the first occurrence of the subsequence, or -1 if not found.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">find_repeated_subsequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">min_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.find_repeated_subsequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Find all subsequences of minimum length that appear more than once.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Set</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>min_length</strong><span class="classifier">int, optional</span></dt><dd><p>Minimum length of subsequence (default is 8).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>set of str</dt><dd><p>Repeated subsequences.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">gc_content</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">extended_alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.gc_content"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Calculate the GC content of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>extended_alphabet</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to include ambiguous base codes (e.g., S, M, R) in the calculation.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>float</dt><dd><p>GC content percentage.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get_random_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.get_random_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Generate a randomized sequence compatible with the IUPAC symbols and
optionally a dot-bracket structure to be respected.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">str, optional</span></dt><dd><p>Dot-bracket structure.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>A new randomized sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">lower</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.lower"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Convert sequence to lowercase.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>str</dt><dd><p>Lowercase sequence string.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">molecular_weight</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.molecular_weight"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Calculate the molecular weight of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>float</dt><dd><p>Total molecular weight in Daltons.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">pop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">idx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.pop"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Remove and return a nucleotide at the given index.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>idx</strong><span class="classifier">int</span></dt><dd><p>Index of nucleotide to remove.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The removed character.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">replace</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">old</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.replace"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Replace all occurrences of <cite>old</cite> with <cite>new</cite> in the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>old</strong><span class="classifier">str</span></dt><dd><p>Character to replace.</p>
</dd>
<dt><strong>new</strong><span class="classifier">str</span></dt><dd><p>Replacement character.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Sequence</dt><dd><p>Updated sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">reverse</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.reverse"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Reverse the directionality of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>inplace</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, reverse in-place. Otherwise, return a new Sequence.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Sequence</dt><dd><p>Reversed sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Return the reverse complement of the sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Sequence</dt><dd><p>Reverse-complemented sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">split</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.split"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Split the sequence by a separator.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>sep</strong><span class="classifier">str, optional</span></dt><dd><p>Separator to use (default is None, meaning split on whitespace).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>list of Sequence</dt><dd><p>List of subsequences.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">translate</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dictionary</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.translate"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Translate the sequence using a mapping dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Sequence</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dictionary</strong><span class="classifier">dict</span></dt><dd><p>Dictionary to translate each character.</p>
</dd>
<dt><strong>inplace</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, modify the current sequence. Otherwise, return a new instance.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Sequence</dt><dd><p>Translated sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">upper</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/sequence.html#Sequence.upper"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Convert sequence to uppercase.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>str</dt><dd><p>Uppercase sequence string.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="basepair">
<h3>Basepair<a class="headerlink" href="#basepair" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.basepair.</span></span><span class="sig-name descname"><span class="pre">BasePair</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">MutableMapping</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">Callback</span></code></p>
<p>A bidirectional dictionary that maintains a mapping between keys and values,
ensuring that each value has a unique corresponding key and vice versa.</p>
<p>Inherits from MutableMapping to provide dictionary-like behavior and from
Callback to support event-driven notifications on updates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>*args</strong><span class="classifier">tuple</span></dt><dd><p>Positional arguments passed to the internal dictionary.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Keyword arguments passed to the internal dictionary. Special keys
<cite>callback</cite> and <cite>callbacks</cite> are removed before processing.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">clear</span></code>()</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>(**kwargs)</p></td>
<td><p>Create a copy of the dictionary.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get</span></code>(key[, default])</p></td>
<td><p>Retrieve a value by key, returning a default if not found.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">items</span></code>()</p></td>
<td><p>Return the dictionary's key-value pairs.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">keys</span></code>()</p></td>
<td><p>Return the dictionary's keys.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>(k[,d])</p></td>
<td><p>If key is not found, d is returned if given, otherwise KeyError is raised.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">popitem</span></code>()</p></td>
<td><p>as a 2-tuple; but raise KeyError if D is empty.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">setdefault</span></code>(k[,d])</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift)</p></td>
<td><p>Apply a shift transformation to the key-value pairs.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">update</span></code>(*args, **kwargs)</p></td>
<td><p>Update the dictionary with new key-value pairs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">values</span></code>()</p></td>
<td><p>Return the dictionary's values.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">copy</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.copy"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Create a copy of the dictionary.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">BasePair</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional parameters for the new instance.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>BasePair</dt><dd><p>A new instance of BasePair with the same data.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">get</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">default</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.get"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Retrieve a value by key, returning a default if not found.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>key</strong><span class="classifier">hashable</span></dt><dd><p>The key to retrieve.</p>
</dd>
<dt><strong>default</strong><span class="classifier">Any, optional</span></dt><dd><p>The value to return if key is not found (default is None).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Any</dt><dd><p>The corresponding value or the default.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">items</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.items"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Return the dictionary’s key-value pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">keys</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.keys"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Return the dictionary’s keys.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></span></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">shift</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.shift"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Apply a shift transformation to the key-value pairs.
The shift is added to both the keys and values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">BasePair</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>shift</strong><span class="classifier">hashable</span></dt><dd><p>The amount to shift each key and value.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>BasePair</dt><dd><p>A new instance with shifted keys and values.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">update</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.update"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Update the dictionary with new key-value pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">BasePair</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>*args</strong><span class="classifier">dict</span></dt><dd><p>Dictionaries to merge into the current instance.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional key-value pairs to update.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>BasePair</dt><dd><p>The updated instance.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py">
<span class="sig-name descname"><span class="pre">values</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/basepair.html#BasePair.values"><span class="viewcode-link"><span class="pre">[source]</span></span></a></dt>
<dd><p>Return the dictionary’s values.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Any</span></code></span></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="common-variables">
<h3>Common Variables<a class="headerlink" href="#common-variables" title="Link to this heading"></a></h3>
<section id="useful-collections">
<h4>Useful collections<a class="headerlink" href="#useful-collections" title="Link to this heading"></a></h4>
<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.nucleotides">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">nucleotides</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'&amp;',</span> <span class="pre">'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'X',</span> <span class="pre">'Y'}</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.nucleotides" title="Link to this definition"></a></dt>
<dd><p>Set of all accepted nucleotides, including standard bases and extended IUPAC codes.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.iupac_code">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">iupac_code</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'&amp;':</span> <span class="pre">{'&amp;'},</span> <span class="pre">'A':</span> <span class="pre">{'A'},</span> <span class="pre">'B':</span> <span class="pre">{'C',</span> <span class="pre">'G',</span> <span class="pre">'U'},</span> <span class="pre">'C':</span> <span class="pre">{'C'},</span> <span class="pre">'D':</span> <span class="pre">{'A',</span> <span class="pre">'G',</span> <span class="pre">'U'},</span> <span class="pre">'G':</span> <span class="pre">{'G'},</span> <span class="pre">'H':</span> <span class="pre">{'A',</span> <span class="pre">'C',</span> <span class="pre">'U'},</span> <span class="pre">'K':</span> <span class="pre">{'G',</span> <span class="pre">'U'},</span> <span class="pre">'M':</span> <span class="pre">{'A',</span> <span class="pre">'C'},</span> <span class="pre">'N':</span> <span class="pre">{'A',</span> <span class="pre">'C',</span> <span class="pre">'G',</span> <span class="pre">'U'},</span> <span class="pre">'R':</span> <span class="pre">{'A',</span> <span class="pre">'G'},</span> <span class="pre">'S':</span> <span class="pre">{'C',</span> <span class="pre">'G'},</span> <span class="pre">'U':</span> <span class="pre">{'U'},</span> <span class="pre">'V':</span> <span class="pre">{'A',</span> <span class="pre">'C',</span> <span class="pre">'G'},</span> <span class="pre">'W':</span> <span class="pre">{'A',</span> <span class="pre">'U'},</span> <span class="pre">'X':</span> <span class="pre">{'A',</span> <span class="pre">'C',</span> <span class="pre">'G',</span> <span class="pre">'U'},</span> <span class="pre">'Y':</span> <span class="pre">{'C',</span> <span class="pre">'U'}}</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.iupac_code" title="Link to this definition"></a></dt>
<dd><p>Dictionary mapping IUPAC codes to corresponding sets of nucleotides.
Source: <a class="reference external" href="https://www.bioinformatics.org/sms/iupac.html">https://www.bioinformatics.org/sms/iupac.html</a></p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.base_pairing">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">base_pairing</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'&amp;':</span> <span class="pre">{'&amp;'},</span> <span class="pre">'A':</span> <span class="pre">{'B',</span> <span class="pre">'D',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'N',</span> <span class="pre">'U',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'B':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'C':</span> <span class="pre">{'B',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'K',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'V'},</span> <span class="pre">'D':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'G':</span> <span class="pre">{'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'H':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'K':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'M':</span> <span class="pre">{'B',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'N':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'R':</span> <span class="pre">{'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'S':</span> <span class="pre">{'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'U':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'V',</span> <span class="pre">'W'},</span> <span class="pre">'V':</span> <span class="pre">{'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'W':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'Y'},</span> <span class="pre">'X':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'X',</span> <span class="pre">'Y'},</span> <span class="pre">'Y':</span> <span class="pre">{'A',</span> <span class="pre">'B',</span> <span class="pre">'D',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'K',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'V',</span> <span class="pre">'W'}}</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.base_pairing" title="Link to this definition"></a></dt>
<dd><p>Dictionary mapping IUPAC codes to the set of codes they can base pair with.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.db_pairs">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">db_pairs</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'(':</span> <span class="pre">')',</span> <span class="pre">'&lt;':</span> <span class="pre">'&gt;',</span> <span class="pre">'A':</span> <span class="pre">'a',</span> <span class="pre">'B':</span> <span class="pre">'b',</span> <span class="pre">'C':</span> <span class="pre">'c',</span> <span class="pre">'D':</span> <span class="pre">'d',</span> <span class="pre">'E':</span> <span class="pre">'e',</span> <span class="pre">'F':</span> <span class="pre">'f',</span> <span class="pre">'G':</span> <span class="pre">'g',</span> <span class="pre">'H':</span> <span class="pre">'h',</span> <span class="pre">'I':</span> <span class="pre">'i',</span> <span class="pre">'J':</span> <span class="pre">'j',</span> <span class="pre">'K':</span> <span class="pre">'k',</span> <span class="pre">'L':</span> <span class="pre">'l',</span> <span class="pre">'M':</span> <span class="pre">'m',</span> <span class="pre">'N':</span> <span class="pre">'n',</span> <span class="pre">'O':</span> <span class="pre">'o',</span> <span class="pre">'P':</span> <span class="pre">'p',</span> <span class="pre">'Q':</span> <span class="pre">'q',</span> <span class="pre">'R':</span> <span class="pre">'r',</span> <span class="pre">'S':</span> <span class="pre">'s',</span> <span class="pre">'T':</span> <span class="pre">'t',</span> <span class="pre">'U':</span> <span class="pre">'u',</span> <span class="pre">'V':</span> <span class="pre">'v',</span> <span class="pre">'W':</span> <span class="pre">'w',</span> <span class="pre">'X':</span> <span class="pre">'x',</span> <span class="pre">'Y':</span> <span class="pre">'y',</span> <span class="pre">'Z':</span> <span class="pre">'z',</span> <span class="pre">'[':</span> <span class="pre">']',</span> <span class="pre">'{':</span> <span class="pre">'}'}</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.db_pairs" title="Link to this definition"></a></dt>
<dd><p>Dictionary mapping dot-bracket symbols to their corresponding pairing symbols.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.all_pk_symbols">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">all_pk_symbols</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">('[',</span> <span class="pre">']',</span> <span class="pre">'{',</span> <span class="pre">'}',</span> <span class="pre">'&lt;',</span> <span class="pre">'&gt;',</span> <span class="pre">'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'J',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'O',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'X',</span> <span class="pre">'Y',</span> <span class="pre">'Z',</span> <span class="pre">'a',</span> <span class="pre">'b',</span> <span class="pre">'c',</span> <span class="pre">'d',</span> <span class="pre">'e',</span> <span class="pre">'f',</span> <span class="pre">'g',</span> <span class="pre">'h',</span> <span class="pre">'i',</span> <span class="pre">'j',</span> <span class="pre">'k',</span> <span class="pre">'l',</span> <span class="pre">'m',</span> <span class="pre">'n',</span> <span class="pre">'o',</span> <span class="pre">'p',</span> <span class="pre">'q',</span> <span class="pre">'r',</span> <span class="pre">'s',</span> <span class="pre">'t',</span> <span class="pre">'u',</span> <span class="pre">'v',</span> <span class="pre">'w',</span> <span class="pre">'x',</span> <span class="pre">'y',</span> <span class="pre">'z')</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.all_pk_symbols" title="Link to this definition"></a></dt>
<dd><p>Tuple containing all possible pseudoknot symbols used in dot-bracket notation.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.accept_symbol">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">accept_symbol</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'</span> <span class="pre">',</span> <span class="pre">'!',</span> <span class="pre">'&amp;',</span> <span class="pre">'(',</span> <span class="pre">')',</span> <span class="pre">'*',</span> <span class="pre">'+',</span> <span class="pre">'-',</span> <span class="pre">'.',</span> <span class="pre">'/',</span> <span class="pre">'3',</span> <span class="pre">'5',</span> <span class="pre">':',</span> <span class="pre">'&lt;',</span> <span class="pre">'=',</span> <span class="pre">'&gt;',</span> <span class="pre">'A',</span> <span class="pre">'B',</span> <span class="pre">'C',</span> <span class="pre">'D',</span> <span class="pre">'E',</span> <span class="pre">'F',</span> <span class="pre">'G',</span> <span class="pre">'H',</span> <span class="pre">'I',</span> <span class="pre">'J',</span> <span class="pre">'K',</span> <span class="pre">'L',</span> <span class="pre">'M',</span> <span class="pre">'N',</span> <span class="pre">'O',</span> <span class="pre">'P',</span> <span class="pre">'Q',</span> <span class="pre">'R',</span> <span class="pre">'S',</span> <span class="pre">'T',</span> <span class="pre">'U',</span> <span class="pre">'V',</span> <span class="pre">'W',</span> <span class="pre">'X',</span> <span class="pre">'Y',</span> <span class="pre">'Z',</span> <span class="pre">'[',</span> <span class="pre">'\\',</span> <span class="pre">']',</span> <span class="pre">'^',</span> <span class="pre">'a',</span> <span class="pre">'b',</span> <span class="pre">'c',</span> <span class="pre">'d',</span> <span class="pre">'e',</span> <span class="pre">'f',</span> <span class="pre">'g',</span> <span class="pre">'h',</span> <span class="pre">'i',</span> <span class="pre">'j',</span> <span class="pre">'k',</span> <span class="pre">'l',</span> <span class="pre">'m',</span> <span class="pre">'n',</span> <span class="pre">'o',</span> <span class="pre">'p',</span> <span class="pre">'q',</span> <span class="pre">'r',</span> <span class="pre">'s',</span> <span class="pre">'t',</span> <span class="pre">'u',</span> <span class="pre">'v',</span> <span class="pre">'w',</span> <span class="pre">'x',</span> <span class="pre">'y',</span> <span class="pre">'z',</span> <span class="pre">'{',</span> <span class="pre">'|',</span> <span class="pre">'}',</span> <span class="pre">'~',</span> <span class="pre">'•',</span> <span class="pre">'↑',</span> <span class="pre">'↓',</span> <span class="pre">'⊗',</span> <span class="pre">'⊙',</span> <span class="pre">'─',</span> <span class="pre">'│',</span> <span class="pre">'┊',</span> <span class="pre">'┼',</span> <span class="pre">'╭',</span> <span class="pre">'╮',</span> <span class="pre">'╯',</span> <span class="pre">'╰',</span> <span class="pre">'▂',</span> <span class="pre">'▄',</span> <span class="pre">'█',</span> <span class="pre">'◦'}</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.accept_symbol" title="Link to this definition"></a></dt>
<dd><p>Set of all accepted symbols including nucleotides, dots, brackets, and special ROAD
symbols.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.bp_symbols">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">bp_symbols</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">{'!',</span> <span class="pre">'*',</span> <span class="pre">':',</span> <span class="pre">'=',</span> <span class="pre">'┊'}</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.bp_symbols" title="Link to this definition"></a></dt>
<dd><p>Set of symbols representing base pairs.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.T7_PROMOTER">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">T7_PROMOTER</span></span><em class="property"><span class="w"> </span><span class="p"><span class="pre">=</span></span><span class="w"> </span><span class="pre">'TAATACGACTCACTATA'</span></em><a class="headerlink" href="#pyfurnace.design.core.symbols.T7_PROMOTER" title="Link to this definition"></a></dt>
<dd><p>Sequence representing the T7 promoter.</p>
</dd></dl>

</section>
<section id="translators-for-strings-characters">
<h4>Translators for strings characters<a class="headerlink" href="#translators-for-strings-characters" title="Link to this heading"></a></h4>
<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.pseudo_to_dot">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">pseudo_to_dot</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.pseudo_to_dot" title="Link to this definition"></a></dt>
<dd><p>String translator mapping all pseudoknot symbols to dots.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.pair_db_sym">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">pair_db_sym</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.pair_db_sym" title="Link to this definition"></a></dt>
<dd><p>String translator to map each dot-bracket opening symbol to its closing counterpart
and vice versa.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.nucl_to_none">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">nucl_to_none</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.nucl_to_none" title="Link to this definition"></a></dt>
<dd><p>String translator that removes all nucleotides from a string.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.symb_to_none">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">symb_to_none</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.symb_to_none" title="Link to this definition"></a></dt>
<dd><p>String translator that removes all accepted symbols from a string.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.nucl_to_pair">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">nucl_to_pair</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.nucl_to_pair" title="Link to this definition"></a></dt>
<dd><p>String translator that maps nucleotides (including IUPAC codes) to their
complementary base.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.only_nucl">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">only_nucl</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.only_nucl" title="Link to this definition"></a></dt>
<dd><p>String translator that removes all non-nucleotide symbols from a string.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.horiz_flip">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">horiz_flip</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.horiz_flip" title="Link to this definition"></a></dt>
<dd><p>String translator for horizontal flip of strand turning symbols.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.verti_flip">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">verti_flip</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.verti_flip" title="Link to this definition"></a></dt>
<dd><p>String translator for vertical flip of strand turning symbols.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.rotate_90">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">rotate_90</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.rotate_90" title="Link to this definition"></a></dt>
<dd><p>String translator for 90 degree rotation of strand symbols.</p>
</dd></dl>

<dl class="py data">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.symb_to_road">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">symb_to_road</span></span><a class="headerlink" href="#pyfurnace.design.core.symbols.symb_to_road" title="Link to this definition"></a></dt>
<dd><p>String translator for converting normal symbols to ROAD strand symbols.</p>
</dd></dl>

</section>
</section>
<section id="module-pyfurnace.design.core.symbols">
<span id="common-functions"></span><h3>Common functions<a class="headerlink" href="#module-pyfurnace.design.core.symbols" title="Link to this heading"></a></h3>
<p>This module contains general utility functions and constants for the pyfurnace package.</p>
<p>It includes:
functions for manipulating RNA strands and sequences,
utilities for RNA structures (dot-bracket notation, pair maps, trees)
folding barriers calculation for canonical co-transcriptional RNA Origami</p>
<dl class="py exception">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.AmbiguosStructure">
<em class="property"><span class="k"><span class="pre">exception</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">AmbiguosStructure</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">message</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#AmbiguosStructure"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.AmbiguosStructure" title="Link to this definition"></a></dt>
<dd><p>Warning raised when a structure is ambiguous or potentially problematic.</p>
<p>This can occur in cases where:
- The strand routing is not clear</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>message</strong><span class="classifier">str</span></dt><dd><p>Description of the ambiguity or issue.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.MotifStructureError">
<em class="property"><span class="k"><span class="pre">exception</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">MotifStructureError</span></span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#MotifStructureError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.MotifStructureError" title="Link to this definition"></a></dt>
<dd><p>Exception raised when a motif’s secondary structure is invalid or inconsistent.</p>
<p>This error typically indicates issues such as:
- Overlapping or conflicting strands
- Joining strands with different directionalities
- Logical inconsistencies in structural motifs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.base_pair_difference">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">base_pair_difference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_map1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_map2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_ind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accept_unpaired_ind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#base_pair_difference"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.base_pair_difference" title="Link to this definition"></a></dt>
<dd><p>Find indices where base pairing differs between two structures.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>s1</strong><span class="classifier">str</span></dt><dd><p>First structure (dot-bracket notation).</p>
</dd>
<dt><strong>s2</strong><span class="classifier">str</span></dt><dd><p>Second structure.</p>
</dd>
<dt><strong>pair_map1</strong><span class="classifier">dict, optional</span></dt><dd><p>Precomputed pair map for s1.</p>
</dd>
<dt><strong>pair_map2</strong><span class="classifier">dict, optional</span></dt><dd><p>Precomputed pair map for s2.</p>
</dd>
<dt><strong>ignore_ind</strong><span class="classifier">iterable of int, optional</span></dt><dd><p>Indices to ignore in the comparison.</p>
</dd>
<dt><strong>accept_unpaired_ind</strong><span class="classifier">iterable of int, optional</span></dt><dd><p>Indices allowed to be unpaired in s2 without penalty.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>list of int</dt><dd><p>Indices of base pairs that differ.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.base_pair_distance">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">base_pair_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_map1</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_map2</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_ind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accept_unpaired_ind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#base_pair_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.base_pair_distance" title="Link to this definition"></a></dt>
<dd><p>Compute the number of base pair differences between two structures.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>s1</strong><span class="classifier">str</span></dt><dd><p>First dot-bracket structure.</p>
</dd>
<dt><strong>s2</strong><span class="classifier">str</span></dt><dd><p>Second dot-bracket structure.</p>
</dd>
<dt><strong>pair_map1</strong><span class="classifier">dict, optional</span></dt><dd><p>Pair map for s1.</p>
</dd>
<dt><strong>pair_map2</strong><span class="classifier">dict, optional</span></dt><dd><p>Pair map for s2.</p>
</dd>
<dt><strong>ignore_ind</strong><span class="classifier">iterable of int, optional</span></dt><dd><p>Indices to ignore.</p>
</dd>
<dt><strong>accept_unpaired_ind</strong><span class="classifier">iterable of int, optional</span></dt><dd><p>Positions where being unpaired in s2 is allowed.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd><p>Total number of mismatched base pairs.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.complement">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.complement" title="Link to this definition"></a></dt>
<dd><p>Return the complement of an RNA sequence using IUPAC base-pairing rules.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str</span></dt><dd><p>RNA sequence composed of standard and extended IUPAC nucleotide symbols.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>Complementary sequence with each base replaced by its pair.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.dot_bracket_to_pair_map">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">dot_bracket_to_pair_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dot_bracket</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#dot_bracket_to_pair_map"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.dot_bracket_to_pair_map" title="Link to this definition"></a></dt>
<dd><p>Convert dot-bracket notation into a pair map.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">BasePair</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dot_bracket</strong><span class="classifier">str</span></dt><dd><p>Secondary structure in dot-bracket notation.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>BasePair</dt><dd><p>A bidirectional dictionary mapping nucleotide indices to their paired
partner (or None).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.dot_bracket_to_stacks">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">dot_bracket_to_stacks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dot_bracket</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_opening</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#dot_bracket_to_stacks"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.dot_bracket_to_stacks" title="Link to this definition"></a></dt>
<dd><p>Identify contiguous structural elements (stacks) in a dot-bracket string.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dot_bracket</strong><span class="classifier">str</span></dt><dd><p>RNA structure in dot-bracket format.</p>
</dd>
<dt><strong>only_opening</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to return only stacks that start with an opening symbol.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>tuple of str and list of tuple</dt><dd><ul class="simple">
<li><p>A reduced dot-bracket string marking distinct stacks.</p></li>
<li><p>A list of (start_index, end_index) tuples for each stack.</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.dot_bracket_to_tree">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">dot_bracket_to_tree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dot_bracket</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#dot_bracket_to_tree"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.dot_bracket_to_tree" title="Link to this definition"></a></dt>
<dd><p>Convert a dot-bracket RNA structure into a hierarchical tree of nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Node</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dot_bracket</strong><span class="classifier">str</span></dt><dd><p>RNA structure in dot-bracket format. Pseudoknots are ignored.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>Sequence constraints associated with each index.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Node</dt><dd><p>Root node of the tree representing the full structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.folding_barriers">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">folding_barriers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kl_delay</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">150</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#folding_barriers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.folding_barriers" title="Link to this definition"></a></dt>
<dd><p>Compute the folding barriers for a given RNA secondary structure.
This function is based on
ROAD: <a class="reference external" href="https://www.nature.com/articles/s41557-021-00679-1">https://www.nature.com/articles/s41557-021-00679-1</a>
This function analyzes the dot-bracket representation of the RNA secondary
structure to determine folding barriers based on base pair topology and
kinetic delay.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">str</span></dt><dd><p>The dot-bracket notation representing the secondary structure of the RNA.
If folding barriers is called from a Motif or Origami, the structure is
already provided by the object in the dot-bracket format.</p>
</dd>
<dt><strong>kl_delay</strong><span class="classifier">int, optional, default=150</span></dt><dd><p>The number of nucleotides (nts) of delay before kissing loops (KLs) snap
closed. A typical realistic value is around 350 (~1 second), while a
more conservative  setting is 150 by default.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Tuple[str, int]</dt><dd><p>A tuple containing:
- A string representing the barrier map where:
‘─’ = no barrier (0 penalty),
‘▂’ = opening pair weak barrier (1 penalty),
‘▄’ = cloding pair weak barrier (1 penalty),
‘█’ = strong barrier (2 penalty).
- An integer score indicating the total penalty based on barrier strengths.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function assigns barrier strengths based on the topology of base pairs
and kinetic constraints, ensuring proper folding predictions.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.gc_content">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">gc_content</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">extended_alphabet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#gc_content"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.gc_content" title="Link to this definition"></a></dt>
<dd><p>Calculate GC content (optionally including ambiguous nucleotides).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>seq</strong><span class="classifier">str</span></dt><dd><p>The RNA sequence.</p>
</dd>
<dt><strong>extended_alphabet</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to include ambiguous IUPAC symbols (e.g., S, K) as partial GC.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>float</dt><dd><p>Proportion of GC content in the sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.hamming_distance">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">hamming_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">s1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_ind</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">()</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#hamming_distance"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.hamming_distance" title="Link to this definition"></a></dt>
<dd><p>Compute the Hamming distance between two strings, ignoring specified indices.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>s1</strong><span class="classifier">str</span></dt><dd><p>First sequence or structure.</p>
</dd>
<dt><strong>s2</strong><span class="classifier">str</span></dt><dd><p>Second sequence or structure.</p>
</dd>
<dt><strong>ignore_ind</strong><span class="classifier">iterable of int, optional</span></dt><dd><p>Indices to exclude from comparison.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>int</dt><dd><p>Number of differing characters at unignored positions.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.mutate_nucleotide">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">mutate_nucleotide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence_constraints</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nucl_ind</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pair_map</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#mutate_nucleotide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.mutate_nucleotide" title="Link to this definition"></a></dt>
<dd><p>Mutate a nucleotide in the sequence and update its paired counterpart accordingly.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>], <code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str</span></dt><dd><p>Original RNA sequence.</p>
</dd>
<dt><strong>sequence_constraints</strong><span class="classifier">str</span></dt><dd><p>A string of IUPAC codes representing base constraints for each position.</p>
</dd>
<dt><strong>nucl_ind</strong><span class="classifier">int</span></dt><dd><p>Index of the nucleotide to mutate.</p>
</dd>
<dt><strong>pair_map</strong><span class="classifier">dict of int to int or None</span></dt><dd><p>A mapping of indices representing base pairs.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>tuple of (str or None, str or None)</dt><dd><p>A tuple with the new nucleotide and its paired base if applicable.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.pair_map_to_dot_bracket">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">pair_map_to_dot_bracket</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pair_map</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure_length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#pair_map_to_dot_bracket"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.pair_map_to_dot_bracket" title="Link to this definition"></a></dt>
<dd><p>Convert a base pair map into dot-bracket notation.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pair_map</strong><span class="classifier">dict</span></dt><dd><p>Mapping from nucleotide index to paired index or None.</p>
</dd>
<dt><strong>structure_length</strong><span class="classifier">int, optional</span></dt><dd><p>Total length of the structure. If None, computed from the map.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>Dot-bracket notation representing the structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.pair_nucleotide">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">pair_nucleotide</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">nucleotide</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iupac_symbol</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'N'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#pair_nucleotide"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.pair_nucleotide" title="Link to this definition"></a></dt>
<dd><p>Return a nucleotide that can pair with the given nucleotide, optionally
constrained by an IUPAC symbol.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>nucleotide</strong><span class="classifier">str</span></dt><dd><p>The base to find a pair for.</p>
</dd>
<dt><strong>iupac_symbol</strong><span class="classifier">str, optional</span></dt><dd><p>Constraint based on allowed base-pairing (IUPAC code).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>A valid pairing nucleotide, or the complement of the input base.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.reverse_complement">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">reverse_complement</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#reverse_complement"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.reverse_complement" title="Link to this definition"></a></dt>
<dd><p>Return the reverse complement of an RNA sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str</span></dt><dd><p>RNA sequence to reverse and complement.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The reverse-complemented sequence.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.rotate_dot_bracket">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">rotate_dot_bracket</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dot_bracket</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift_left</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#rotate_dot_bracket"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.rotate_dot_bracket" title="Link to this definition"></a></dt>
<dd><p>Rotate a dot-bracket structure leftward by a given number of positions.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dot_bracket</strong><span class="classifier">str</span></dt><dd><p>RNA secondary structure in dot-bracket notation.</p>
</dd>
<dt><strong>shift_left</strong><span class="classifier">int</span></dt><dd><p>Number of positions to rotate to the left.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>Rotated dot-bracket structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.core.symbols.tree_to_dot_bracket">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.core.symbols.</span></span><span class="sig-name descname"><span class="pre">tree_to_dot_bracket</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dot_bracket</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_constraints</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/core/symbols.html#tree_to_dot_bracket"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.core.symbols.tree_to_dot_bracket" title="Link to this definition"></a></dt>
<dd><p>Convert a tree of Nodes into a dot-bracket structure and optional sequence
constraints.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>node</strong><span class="classifier">Node</span></dt><dd><p>Root node of the RNA structure tree.</p>
</dd>
<dt><strong>dot_bracket</strong><span class="classifier">list of str, optional</span></dt><dd><p>Accumulator for the dot-bracket string.</p>
</dd>
<dt><strong>seq_constraints</strong><span class="classifier">bool or list of str</span></dt><dd><p>If True, collect sequence constraints using the <cite>seq</cite> attribute.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str or tuple of (str, str)</dt><dd><p>Dot-bracket string, and sequence constraints if requested.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="motifs">
<h2>Motifs<a class="headerlink" href="#motifs" title="Link to this heading"></a></h2>
<section id="module-pyfurnace.design.motifs.stem">
<span id="stem"></span><h3>Stem<a class="headerlink" href="#module-pyfurnace.design.motifs.stem" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.stem.</span></span><span class="sig-name descname"><span class="pre">Stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wobble_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wobble_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wobble_insert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'middle'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strong_bases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/stem.html#Stem"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem" title="Link to this definition"></a></dt>
<dd><p>Represents a helical RNA stem motif consisting of complementary base-paired strands.</p>
<p>The <cite>Stem</cite> class models a double-stranded RNA region with optional support for
wobble base pairs and short-sequences set to strong bases.
It can be created from a user-defined sequence or automatically
generated based on a target length.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>length</strong><span class="classifier">int, optional</span></dt><dd><p>Number of base pairs in the stem. Ignored if <cite>sequence</cite> is provided.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str or Sequence, optional</span></dt><dd><p>Sequence for the top strand. If provided, the bottom strand is inferred
as its reverse complement.</p>
</dd>
<dt><strong>wobble_interval</strong><span class="classifier">int, optional</span></dt><dd><p>Spacing interval for inserting wobble base pairs, e.g., every N bases.
Default is 5. Ignored if <cite>sequence</cite> is given.</p>
</dd>
<dt><strong>wobble_tolerance</strong><span class="classifier">int, optional</span></dt><dd><p>Maximum random deviation from the defined wobble interval. Default is 2.</p>
</dd>
<dt><strong>wobble_insert</strong><span class="classifier">{“middle”, “start”, “end”}, optional</span></dt><dd><p>Positioning strategy for wobble base insertions. Default is “middle”.</p>
</dd>
<dt><strong>strong_bases</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, use strong (GC) base pairing for short stems (≤ 3 bp).
Default is True.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict, optional</span></dt><dd><p>Additional keyword arguments passed to the <cite>Motif</cite> superclass.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#pyfurnace.design.motifs.stem.Stem.length" title="pyfurnace.design.motifs.stem.Stem.length"><code class="xref py py-obj docutils literal notranslate"><span class="pre">length</span></code></a><span class="classifier">int</span></dt><dd><p>Number of nucleotides in a stem</p>
</dd>
<dt><strong>wobble_interval</strong><span class="classifier">int</span></dt><dd><p>Interval used for wobble base insertion. Zero if sequence is specified.</p>
</dd>
<dt><strong>wobble_tolerance</strong><span class="classifier">int</span></dt><dd><p>Allowed range of randomness around the wobble interval.</p>
</dd>
<dt><strong>wobble_insert</strong><span class="classifier">str</span></dt><dd><p>Strategy for where to place wobble base pairs.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code><span class="classifier">list of Strand</span></dt><dd><p>The list of strands in the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.stem.Stem.set_down_sequence" title="pyfurnace.design.motifs.stem.Stem.set_down_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_down_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the bottom strand</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.stem.Stem.set_strong_bases" title="pyfurnace.design.motifs.stem.Stem.set_strong_bases"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_strong_bases</span></code></a>(strong_bases)</p></td>
<td><p>Set wether to use strong bases for short stems</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.stem.Stem.set_up_sequence" title="pyfurnace.design.motifs.stem.Stem.set_up_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_up_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the top strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.length">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.length" title="Link to this definition"></a></dt>
<dd><p>Number of nucleotides in a stem</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.set_down_sequence">
<span class="sig-name descname"><span class="pre">set_down_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/stem.html#Stem.set_down_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.set_down_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the bottom strand</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.set_strong_bases">
<span class="sig-name descname"><span class="pre">set_strong_bases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strong_bases</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/stem.html#Stem.set_strong_bases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.set_strong_bases" title="Link to this definition"></a></dt>
<dd><p>Set wether to use strong bases for short stems</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.set_up_sequence">
<span class="sig-name descname"><span class="pre">set_up_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/stem.html#Stem.set_up_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.set_up_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the top strand</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.wobble_insert">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wobble_insert</span></span><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.wobble_insert" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.wobble_interval">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wobble_interval</span></span><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.wobble_interval" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.stem.Stem.wobble_tolerance">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">wobble_tolerance</span></span><a class="headerlink" href="#pyfurnace.design.motifs.stem.Stem.wobble_tolerance" title="Link to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

</section>
<section id="module-pyfurnace.design.motifs.dovetail">
<span id="dovetail"></span><h3>Dovetail<a class="headerlink" href="#module-pyfurnace.design.motifs.dovetail" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.dovetail.Dovetail">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.dovetail.</span></span><span class="sig-name descname"><span class="pre">Dovetail</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">up_cross</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">down_cross</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wobble_interval</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wobble_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wobble_insert</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'middle'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strong_bases</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/dovetail.html#Dovetail"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.dovetail.Dovetail" title="Link to this definition"></a></dt>
<dd><p>Represents a double helix RNA stem with junction crossovers before and after
the stem.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>length</strong><span class="classifier">int, optional</span></dt><dd><p>Number of base pairs in the stem. The sign determines dovetail direction.
Ignored if <cite>sequence</cite> is provided.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>RNA sequence to assign to the top strand. Overrides <cite>length</cite>.</p>
</dd>
<dt><strong>up_cross</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, includes a top dovetail crossover motif. Default is True.</p>
</dd>
<dt><strong>down_cross</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, includes a bottom dovetail crossover motif. Default is True.</p>
</dd>
<dt><strong>sign</strong><span class="classifier">int, optional</span></dt><dd><p>Direction of the dovetail: +1 (positive/right) or -1 (negative/left).
If 0 or unspecified, inferred from <cite>length</cite> or <cite>sequence</cite>. Default is 1.</p>
</dd>
<dt><strong>wobble_interval</strong><span class="classifier">int, optional</span></dt><dd><p>Periodicity of wobble base pairs in the stem. Default is 5.</p>
</dd>
<dt><strong>wobble_tolerance</strong><span class="classifier">int, optional</span></dt><dd><p>Allowed deviation in wobble periodicity. Default is 2.</p>
</dd>
<dt><strong>wobble_insert</strong><span class="classifier">str, optional</span></dt><dd><p>Position of wobble insertions. Must be “start”, “middle”, or “end”.
Default is “middle”.</p>
</dd>
<dt><strong>strong_bases</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to enforce GC-rich base pairs. If None, it defaults to True only
when both <cite>up_cross</cite> and <cite>down_cross</cite> are True.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments passed to the parent <cite>Stem</cite> class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#pyfurnace.design.motifs.dovetail.Dovetail.up_cross" title="pyfurnace.design.motifs.dovetail.Dovetail.up_cross"><code class="xref py py-obj docutils literal notranslate"><span class="pre">up_cross</span></code></a><span class="classifier">bool</span></dt><dd><p>Returns boolian describing wether the dovetail has a top crossing</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.motifs.dovetail.Dovetail.down_cross" title="pyfurnace.design.motifs.dovetail.Dovetail.down_cross"><code class="xref py py-obj docutils literal notranslate"><span class="pre">down_cross</span></code></a><span class="classifier">bool</span></dt><dd><p>Returns boolian describing wether the dovetail has a bottom crossing</p>
</dd>
<dt><strong>sign</strong><span class="classifier">int</span></dt><dd><p>Direction of the dovetail: +1 (right/positive), -1 (left/negative).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">length</span></code><span class="classifier">int</span></dt><dd><p>Number of nucleotides in a stem</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code><span class="classifier">str</span></dt><dd><p>Return the sequence of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.dovetail.Dovetail.set_down_sequence" title="pyfurnace.design.motifs.dovetail.Dovetail.set_down_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_down_sequence</span></code></a>(sequence[, sign])</p></td>
<td><p>Set the sequence of the bottom strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_strong_bases</span></code>(strong_bases)</p></td>
<td><p>Set wether to use strong bases for short stems</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.dovetail.Dovetail.set_up_sequence" title="pyfurnace.design.motifs.dovetail.Dovetail.set_up_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_up_sequence</span></code></a>(sequence[, sign])</p></td>
<td><p>Set the sequence of the top strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.dovetail.Dovetail.down_cross">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">down_cross</span></span><a class="headerlink" href="#pyfurnace.design.motifs.dovetail.Dovetail.down_cross" title="Link to this definition"></a></dt>
<dd><p>Returns boolian describing wether the dovetail has a bottom crossing</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.dovetail.Dovetail.set_down_sequence">
<span class="sig-name descname"><span class="pre">set_down_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/dovetail.html#Dovetail.set_down_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.dovetail.Dovetail.set_down_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the bottom strand</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.dovetail.Dovetail.set_up_sequence">
<span class="sig-name descname"><span class="pre">set_up_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/dovetail.html#Dovetail.set_up_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.dovetail.Dovetail.set_up_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the top strand</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.dovetail.Dovetail.up_cross">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">up_cross</span></span><a class="headerlink" href="#pyfurnace.design.motifs.dovetail.Dovetail.up_cross" title="Link to this definition"></a></dt>
<dd><p>Returns boolian describing wether the dovetail has a top crossing</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pyfurnace.design.motifs.loops">
<span id="loops"></span><h3>Loops<a class="headerlink" href="#module-pyfurnace.design.motifs.loops" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.loops.Loop">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.loops.</span></span><span class="sig-name descname"><span class="pre">Loop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/loops.html#Loop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.loops.Loop" title="Link to this definition"></a></dt>
<dd><p>Represents a generic unpaired loop motif in an RNA structure.
Optionally it can be initialized with a sequence or be oriented
to the left.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>open_left</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to flip the loop horizontally and vertically to open to the left.
Default is False.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>Nucleotide sequence to include in the loop. If provided, creates a strand
with the sequence and loop symbols. Default is “”.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict, optional</span></dt><dd><p>Additional keyword arguments passed to the parent <cite>Motif</cite> class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code><span class="classifier">list of Strand</span></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><strong>flipped</strong><span class="classifier">bool</span></dt><dd><p>Whether the motif has been flipped horizontally/vertically
(if <cite>open_left=True</cite>).</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.loops.TetraLoop">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.loops.</span></span><span class="sig-name descname"><span class="pre">TetraLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'UUCG'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/loops.html#TetraLoop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.loops.TetraLoop" title="Link to this definition"></a></dt>
<dd><p>Represents a specific 4-nucleotide tetraloop RNA motif.
This class implements a canonical tetraloop structure (e.g., UUCG) with a
predefined folding pattern and 3D coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>open_left</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to flip the loop to open to the left. Default is False.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>4-nucleotide RNA sequence for the tetraloop. Default is “UUCG”.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict, optional</span></dt><dd><p>Additional keyword arguments passed to the <cite>Loop</cite> superclass. You can
override the default strand using the <cite>strands</cite> argument.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code><span class="classifier">list of Strand</span></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code><span class="classifier">str</span></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><strong>_coords</strong><span class="classifier">Coords</span></dt><dd><p>3D coordinate data loaded from template structure (e.g., from PDB 2KOC).</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.loops.TetraLoop.set_sequence" title="pyfurnace.design.motifs.loops.TetraLoop.set_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code></a>(new_sequence)</p></td>
<td><p>Set a new 4-nucleotide sequence for the tetraloop.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.loops.TetraLoop.set_sequence">
<span class="sig-name descname"><span class="pre">set_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_sequence</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/loops.html#TetraLoop.set_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.loops.TetraLoop.set_sequence" title="Link to this definition"></a></dt>
<dd><p>Set a new 4-nucleotide sequence for the tetraloop.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>new_sequence</strong><span class="classifier">str</span></dt><dd><p>New tetraloop sequence. Must be exactly 4 nucleotides.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>None</dt><dd></dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the provided sequence is not exactly 4 nucleotides long.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>
<section id="module-pyfurnace.design.motifs.kissing_loops">
<span id="kissing-loops"></span><h3>Kissing loops<a class="headerlink" href="#module-pyfurnace.design.motifs.kissing_loops" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.BranchedDimer">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">BranchedDimer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#BranchedDimer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.BranchedDimer" title="Link to this definition"></a></dt>
<dd><p>Branched kissing loop dimer with top and bottom interactions.
The motif consists of three strands:
- 1) A 180°KL dimer strand
- 2) A branched kissing loop strand
- 3) A connection strand for the branched kissing loop</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>Sequence for the top kissing strand. Default is “”.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Identifier for the pseudoknot. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Free energy of the motif. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Energy tolerance threshold. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_down_sequence" title="pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_down_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_down_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the bottom strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_up_sequence" title="pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_up_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_up_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the top strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_down_sequence">
<span class="sig-name descname"><span class="pre">set_down_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#BranchedDimer.set_down_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_down_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the bottom strand</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_up_sequence">
<span class="sig-name descname"><span class="pre">set_up_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#BranchedDimer.set_up_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.BranchedDimer.set_up_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the top strand</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">BranchedKissingLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#BranchedKissingLoop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop" title="Link to this definition"></a></dt>
<dd><p>Represents a 180° kissing loop motif with a branch connection.
The motif consists of two strands:
- 1) A kissing loop strand with a branch
- 2) A connection strand for the branched kissing loop</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>open_left</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether the motif is open on the left side. Default is False.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>RNA sequence used for the kissing strand. Default is an empty string.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Pseudoknot identifier. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Free energy of the structure. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Tolerance for energy variations. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop.get_kissing_sequence" title="pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop.get_kissing_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code></a>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop.get_kissing_sequence">
<span class="sig-name descname"><span class="pre">get_kissing_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#BranchedKissingLoop.get_kissing_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.BranchedKissingLoop.get_kissing_sequence" title="Link to this definition"></a></dt>
<dd><p>Returns the kissing sequence of the kissing loop</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingDimer">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">KissingDimer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingDimer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer" title="Link to this definition"></a></dt>
<dd><p>A kissing loop dimer composed of two complementary 180-degree loops.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>Sequence of the top strand; bottom is inferred as reverse complement.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Pseudoknot identifier for the dimer. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Free energy of the dimer. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Tolerance for energy deviation. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer.set_down_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingDimer.set_down_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_down_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the bottom strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer.set_up_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingDimer.set_up_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_up_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the top strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingDimer.set_down_sequence">
<span class="sig-name descname"><span class="pre">set_down_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingDimer.set_down_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer.set_down_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the bottom strand</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingDimer.set_up_sequence">
<span class="sig-name descname"><span class="pre">set_up_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingDimer.set_up_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer.set_up_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the top strand</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingDimer120">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">KissingDimer120</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingDimer120"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer120" title="Link to this definition"></a></dt>
<dd><p>Dimer of two complementary kissing loops based on the 120-degree model.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>RNA sequence for the top strand. Default is “”.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Pseudoknot symbol. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Energy of the dimer motif. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Allowed energy deviation. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_down_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_down_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_down_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the bottom strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_up_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_up_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_up_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the top strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_down_sequence">
<span class="sig-name descname"><span class="pre">set_down_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingDimer120.set_down_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_down_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the bottom strand</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_up_sequence">
<span class="sig-name descname"><span class="pre">set_up_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingDimer120.set_up_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingDimer120.set_up_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the top strand</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">KissingLoop</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop" title="Link to this definition"></a></dt>
<dd><p>Represents a pseudoknotted kissing loop motif in RNA secondary structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>open_left</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, the loop is open on the left side. Default is False.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>RNA sequence of the internal kissing loop. Default is an empty string.</p>
</dd>
<dt><strong>seq_len</strong><span class="classifier">int, optional</span></dt><dd><p>Expected length of the sequence. Used for validation. Default is 0.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Pseudoknot index or symbol identifying the loop. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Free energy (in kcal/mol) associated with the loop. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Tolerance on the energy for acceptable structural variants. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments passed to the parent <cite>Loop</cite> constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.pk_index" title="pyfurnace.design.motifs.kissing_loops.KissingLoop.pk_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></a><span class="classifier">str</span></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.energy" title="pyfurnace.design.motifs.kissing_loops.KissingLoop.energy"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></a><span class="classifier">float</span></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.energy_tolerance" title="pyfurnace.design.motifs.kissing_loops.KissingLoop.energy_tolerance"><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></a><span class="classifier">float</span></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.complementary_pk_index" title="pyfurnace.design.motifs.kissing_loops.KissingLoop.complementary_pk_index"><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code></a>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.get_kissing_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingLoop.get_kissing_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code></a>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.set_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingLoop.set_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code></a>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop.complementary_pk_index">
<span class="sig-name descname"><span class="pre">complementary_pk_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop.complementary_pk_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.complementary_pk_index" title="Link to this definition"></a></dt>
<dd><p>Returns the complementary pseudoknot symbol of the kissing loop</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>The pseudoknot index to complement. If None, uses the current pk_index.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>The complementary pseudoknot index.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop.energy">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy</span></span><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.energy" title="Link to this definition"></a></dt>
<dd><p>Returns the energy of the internal kissing loop</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop.energy_tolerance">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">energy_tolerance</span></span><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.energy_tolerance" title="Link to this definition"></a></dt>
<dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop.get_kissing_sequence">
<span class="sig-name descname"><span class="pre">get_kissing_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop.get_kissing_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.get_kissing_sequence" title="Link to this definition"></a></dt>
<dd><p>Returns the kissing sequence of the kissing loop</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop.pk_index">
<em class="property"><span class="k"><span class="pre">property</span></span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">pk_index</span></span><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.pk_index" title="Link to this definition"></a></dt>
<dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop.set_sequence">
<span class="sig-name descname"><span class="pre">set_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">new_seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop.set_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop.set_sequence" title="Link to this definition"></a></dt>
<dd><p>Set the sequence of the strand</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop120">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">KissingLoop120</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop120"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop120" title="Link to this definition"></a></dt>
<dd><p>Kissing loop motif based on structural template from PDB entry 1BJ2.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>open_left</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether the loop is open on the left side. Default is False.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>RNA sequence to use. Default is an empty string.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Pseudoknot index or label. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Free energy (kcal/mol) of the loop. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Tolerance on the energy value. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional arguments passed to the parent constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop180">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.kissing_loops.</span></span><span class="sig-name descname"><span class="pre">KissingLoop180</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pk_index</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">energy_tolerance</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop180"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop180" title="Link to this definition"></a></dt>
<dd><p>Kissing loop from HIV with idealized helical geometry for 180-degree interaction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>open_left</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether the loop is open on the left. Default is False.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str, optional</span></dt><dd><p>RNA sequence for the motif. Default is an empty string.</p>
</dd>
<dt><strong>pk_index</strong><span class="classifier">str or int, optional</span></dt><dd><p>Pseudoknot index for identification. Default is ‘0’.</p>
</dd>
<dt><strong>energy</strong><span class="classifier">float, optional</span></dt><dd><p>Free energy of the motif. Default is -8.5.</p>
</dd>
<dt><strong>energy_tolerance</strong><span class="classifier">float, optional</span></dt><dd><p>Acceptable energy deviation. Default is 1.0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy</span></code></dt><dd><p>Returns the energy of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">energy_tolerance</span></code></dt><dd><p>Returns the energy tolerance of the internal kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pk_index</span></code></dt><dd><p>Returns the pseudoknot symbol of the kissing loop</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">complementary_pk_index</span></code>([pk_index])</p></td>
<td><p>Returns the complementary pseudoknot symbol of the kissing loop</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop180.get_kissing_sequence" title="pyfurnace.design.motifs.kissing_loops.KissingLoop180.get_kissing_sequence"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_kissing_sequence</span></code></a>()</p></td>
<td><p>Returns the kissing sequence of the kissing loop</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">set_sequence</span></code>(new_seq)</p></td>
<td><p>Set the sequence of the strand</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.kissing_loops.KissingLoop180.get_kissing_sequence">
<span class="sig-name descname"><span class="pre">get_kissing_sequence</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/kissing_loops.html#KissingLoop180.get_kissing_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.kissing_loops.KissingLoop180.get_kissing_sequence" title="Link to this definition"></a></dt>
<dd><p>Returns the kissing sequence of the kissing loop</p>
</dd></dl>

</dd></dl>

</section>
<section id="module-pyfurnace.design.motifs.aptamers">
<span id="aptamers"></span><h3>Aptamers<a class="headerlink" href="#module-pyfurnace.design.motifs.aptamers" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Aptamer">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Aptamer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">strands</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basepair</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">structure</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">autopairing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lock_coords</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">join</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">copy</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Aptamer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Aptamer" title="Link to this definition"></a></dt>
<dd><p>Base class for all aptamers, inheriting from the Motif class.
This class is used to create aptamers with specific sequences and coordinates.
At the moment, it does not add any additional functionality to the Motif class,
but it can be used to create a common interface for all aptamers or for screening
motifs in an origami.</p>
<dl class="field-list simple">
<dt class="field-odd">Attributes<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Biotin">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Biotin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Biotin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Biotin" title="Link to this definition"></a></dt>
<dd><p>Returns the biotin-binding aptamer motif.</p>
<p>Based on PDB structure 1F27.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Broccoli">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Broccoli</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Broccoli"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Broccoli" title="Link to this definition"></a></dt>
<dd><p>Returns the Broccoli aptamer motif.</p>
<p>A fluorescent aptamer derived from PDB 7ZJ5.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Ispinach">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Ispinach</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Ispinach"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Ispinach" title="Link to this definition"></a></dt>
<dd><p>Returns the ISpinach aptamer motif.</p>
<p>A fluorescent RNA aptamer modeled from PDB 5OB3.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.L7Ae">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">L7Ae</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#L7Ae"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.L7Ae" title="Link to this definition"></a></dt>
<dd><p>Returns the L7Ae protein-binding aptamer.</p>
<p>Structure based on PDB 1RLG.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.MS2">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">MS2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#MS2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.MS2" title="Link to this definition"></a></dt>
<dd><p>Returns the MS2 coat protein-binding aptamer.</p>
<p>Based on PDB 1ZDH, extended by 3 nucleotides.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.MalachiteGreen">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">MalachiteGreen</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#MalachiteGreen"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.MalachiteGreen" title="Link to this definition"></a></dt>
<dd><p>Returns the full Malachite Green aptamer motif.</p>
<p>Derived from the structure in PDB 1Q8N.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.MalachiteGreenShort">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">MalachiteGreenShort</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#MalachiteGreenShort"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.MalachiteGreenShort" title="Link to this definition"></a></dt>
<dd><p>Returns a truncated version of the Malachite Green aptamer.</p>
<p>Designed for minimal structural features. Based on PDB 1Q8N.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Mango">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Mango</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Mango"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Mango" title="Link to this definition"></a></dt>
<dd><p>Returns the Mango aptamer as a loop-based motif.</p>
<p>A fluorogenic aptamer based on PDB 5V3F.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.PIP3">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">PIP3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#PIP3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.PIP3" title="Link to this definition"></a></dt>
<dd><p>Returns the PIP3-binding aptamer motif.
Pubblication: <a class="reference external" href="https://doi.org/10.1038/ncb3473">https://doi.org/10.1038/ncb3473</a></p>
<p>Structure generated with RNA Composer for synthetic design.
(<a class="reference external" href="https://doi.org/10.1093/nar/gks339">https://doi.org/10.1093/nar/gks339</a>, <a class="reference external" href="https://doi.org/10.1002/prot.26578">https://doi.org/10.1002/prot.26578</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.PIP3_mut1">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">PIP3_mut1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#PIP3_mut1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.PIP3_mut1" title="Link to this definition"></a></dt>
<dd><p>Returns mutant 1 of the PIP3 aptamer.
Sequence-modified variant for functional screening.
Pubblication: <a class="reference external" href="https://doi.org/10.1038/ncb3473">https://doi.org/10.1038/ncb3473</a></p>
<p>Structure generated with RNA Composer for synthetic design.
(<a class="reference external" href="https://doi.org/10.1093/nar/gks339">https://doi.org/10.1093/nar/gks339</a>, <a class="reference external" href="https://doi.org/10.1002/prot.26578">https://doi.org/10.1002/prot.26578</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.PIP3_mut3">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">PIP3_mut3</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#PIP3_mut3"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.PIP3_mut3" title="Link to this definition"></a></dt>
<dd><p>Returns mutant 3 of the PIP3 aptamer.
Sequence-modified variant for functional screening.
Pubblication: <a class="reference external" href="https://doi.org/10.1038/ncb3473">https://doi.org/10.1038/ncb3473</a></p>
<p>Structure generated with RNA Composer for synthetic design.
(<a class="reference external" href="https://doi.org/10.1093/nar/gks339">https://doi.org/10.1093/nar/gks339</a>, <a class="reference external" href="https://doi.org/10.1002/prot.26578">https://doi.org/10.1002/prot.26578</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.PIP3_mut5">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">PIP3_mut5</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#PIP3_mut5"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.PIP3_mut5" title="Link to this definition"></a></dt>
<dd><p>Returns mutant 5 of the PIP3 aptamer.
Sequence-modified variant for functional screening.
Pubblication: <a class="reference external" href="https://doi.org/10.1038/ncb3473">https://doi.org/10.1038/ncb3473</a></p>
<p>Structure generated with RNA Composer for synthetic design.
(<a class="reference external" href="https://doi.org/10.1093/nar/gks339">https://doi.org/10.1093/nar/gks339</a>, <a class="reference external" href="https://doi.org/10.1002/prot.26578">https://doi.org/10.1002/prot.26578</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.PP7">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">PP7</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#PP7"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.PP7" title="Link to this definition"></a></dt>
<dd><p>Returns the PP7 coat protein-binding aptamer.</p>
<p>Derived from PDB 2QUX.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Pepper">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Pepper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Pepper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Pepper" title="Link to this definition"></a></dt>
<dd><p>Returns the Pepper aptamer motif.</p>
<p>A compact fluorescent RNA aptamer. Based on PDB 7ZJ5.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Streptavidin">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Streptavidin</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Streptavidin"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Streptavidin" title="Link to this definition"></a></dt>
<dd><p>Returns the Streptavidin-binding aptamer.
Pubblication: <a class="reference external" href="https://doi.org/10.1093/nar/gkt956">https://doi.org/10.1093/nar/gkt956</a></p>
<p>Modeled using RNA Composer based on experimental data.
(<a class="reference external" href="https://doi.org/10.1093/nar/gks339">https://doi.org/10.1093/nar/gks339</a>, <a class="reference external" href="https://doi.org/10.1002/prot.26578">https://doi.org/10.1002/prot.26578</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.TAR_TAT">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">TAR_TAT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#TAR_TAT"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.TAR_TAT" title="Link to this definition"></a></dt>
<dd><p>Returns the HIV TAR-TAT binding aptamer.</p>
<p>Structure based on PDB 6MCE.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Thrombin_exosite1">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Thrombin_exosite1</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Thrombin_exosite1"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Thrombin_exosite1" title="Link to this definition"></a></dt>
<dd><p>Returns the thrombin exosite 1-binding aptamer.
Used for anticoagulant and structural studies.
Pubblication: <a class="reference external" href="https://doi.org/10.1111/j.1538-7836.2012.04679.x">https://doi.org/10.1111/j.1538-7836.2012.04679.x</a></p>
<p>Modeled using RNA Composer based on experimental data.
(<a class="reference external" href="https://doi.org/10.1093/nar/gks339">https://doi.org/10.1093/nar/gks339</a>, <a class="reference external" href="https://doi.org/10.1002/prot.26578">https://doi.org/10.1002/prot.26578</a>)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.Thrombin_exosite2">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">Thrombin_exosite2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">open_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#Thrombin_exosite2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.Thrombin_exosite2" title="Link to this definition"></a></dt>
<dd><p>Returns the thrombin exosite 2-binding aptamer.</p>
<p>Based on modified PDB structure 5DO4
(replaced fluorinated ribose with regular ribose).</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.motifs.aptamers.create_aptamer">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.motifs.aptamers.</span></span><span class="sig-name descname"><span class="pre">create_aptamer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inherit_from</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/motifs/aptamers.html#create_aptamer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.motifs.aptamers.create_aptamer" title="Link to this definition"></a></dt>
<dd><p>Create an Aptamer object, optionally inheriting from a specified base class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>args</strong><span class="classifier">list</span></dt><dd><p>Positional arguments to be passed to the Aptamer constructor.</p>
</dd>
<dt><strong>inherit_from</strong><span class="classifier">classmethod, optional</span></dt><dd><p>A class from which the Aptamer class should inherit. If not provided,
the Aptamer class will be used as is.</p>
</dd>
<dt><strong>kwargs</strong><span class="classifier">dict</span></dt><dd><p>Keyword arguments to be passed to the Aptamer constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Aptamer</dt><dd><dl class="simple">
<dt>An instance of the Aptamer class, optionally inheriting from the specified</dt><dd><p>inherit_from class.</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="utilities">
<h2>Utilities<a class="headerlink" href="#utilities" title="Link to this heading"></a></h2>
<section id="module-pyfurnace.design.utils.motif_lib">
<span id="motif-utilities"></span><h3>Motif Utilities<a class="headerlink" href="#module-pyfurnace.design.utils.motif_lib" title="Link to this heading"></a></h3>
<dl class="py class">
<dt class="sig sig-object py" id="pyfurnace.design.utils.motif_lib.Utils">
<em class="property"><span class="k"><span class="pre">class</span></span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.motif_lib.</span></span><span class="sig-name descname"><span class="pre">Utils</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hflip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vflip</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rotate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/motif_lib.html#Utils"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.motif_lib.Utils" title="Link to this definition"></a></dt>
<dd><p>Utility class that extends the Motif class with optional flipping and rotation.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>*args</strong><span class="classifier">tuple</span></dt><dd><p>Positional arguments passed to the base <cite>Motif</cite> class.</p>
</dd>
<dt><strong>hflip</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, apply a horizontal flip to the motif. Default is False.</p>
</dd>
<dt><strong>vflip</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, apply a vertical flip to the motif. Default is False.</p>
</dd>
<dt><strong>rotate</strong><span class="classifier">int, optional</span></dt><dd><p>Rotation in degrees (usually 90, 180, 270). Default is 0.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments passed to the <cite>Motif</cite> base class.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Attributes<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">autopairing</span></code></dt><dd><p>Indicates whether the motif automatically pairs bases.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">basepair</span></code></dt><dd><p>A dictionary with positions as key and the paired position as values.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">callbacks</span></code></dt><dd><p>Get the list of registered callbacks.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">junctions</span></code></dt><dd><p>A dictionary with Direction as key (right: (1,0), bottom: (0,1), left: (-1,0), top: (0,1)) and an ordered list of junction positon as value.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">lock_coords</span></code></dt><dd><p>Boolean indicating if the coordinates are locked in the same block.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">max_pos</span></code></dt><dd><p>The maximum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">min_pos</span></code></dt><dd><p>The minimum x, y coordinates occupied of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_char</span></code></dt><dd><p>The maximum length of the motif lines.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">num_lines</span></code></dt><dd><p>The number of lines in the motif structure.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">pair_map</span></code></dt><dd><p>The dictionary of the paired indexes (alternative to the dot bracket notation).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">positions</span></code></dt><dd><p>A tuple of the positions of each character of each strand in 2D space (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">seq_positions</span></code></dt><dd><p>The positions of each nucleotide in the motif sequence (x,y coordinates).</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">sequence</span></code></dt><dd><p>Return the sequence of the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">strands</span></code></dt><dd><p>The list of strands in the motif.</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">structure</span></code></dt><dd><p>Return the dot bracket representation of the motif.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<tbody>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">align</span></code>(*motifs[, axis, extend, ...])</p></td>
<td><p>Align motifs along a given axis by shifting them (without concatenating).</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">append</span></code>(strand[, join, copy])</p></td>
<td><p>Append a strand to the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">concat</span></code>(*motifs[, axis, extend, copy, align, ...])</p></td>
<td><p>Concatenate multiple motifs along a specified axis.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy</span></code>([callback])</p></td>
<td><p>Return a copy of the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">copy_strands_preserve_blocks</span></code>(strands[, motif])</p></td>
<td><p>Copy a list of strands while preserving strand blocks.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">extend_junctions</span></code>([skip_axis, ...])</p></td>
<td><p>Extend the junctions of the motif in the direction of the junctions.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">flip</span></code>([horizontally, vertically, ...])</p></td>
<td><p>Flip the strands of the motif horizontally and/or vertically inplace.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">folding_barriers</span></code>([kl_delay])</p></td>
<td><p>Compute the folding barriers for a given RNA secondary structure.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_file</span></code>(file_path, **kwargs)</p></td>
<td><p>Create a Motif object from a text file containing a motif sketch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_list</span></code>(motif_list, **kwargs)</p></td>
<td><p>Create a Motif object from a list of strings representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_structure</span></code>([structure, sequence, ...])</p></td>
<td><p>Parse a structure or sequence representation to a Motif object.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">from_text</span></code>(motif_text, **kwargs)</p></td>
<td><p>Create a Motif object from a text string representing a motif sketch.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_position_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the corresponding characters as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_sequential_shift</span></code>(motifs, *args[, axis, ...])</p></td>
<td><p>Calculate the shift needed to align motifs sequentially.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_strand_index_map</span></code>()</p></td>
<td><p>Get a dictionary of positions as keys and the index of the strand in the motif as values.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">insert</span></code>(index, strand[, join, copy])</p></td>
<td><p>Insert a strand at a given index in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">join_strands</span></code>(strands)</p></td>
<td><p>Attempt to join consecutive strands and return the list of joined strands.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">pop</span></code>([index])</p></td>
<td><p>Remove and return the strand at the specified index.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">register_callback</span></code>(callback)</p></td>
<td><p>Register a new callback function to be executed when an event occurs.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">replace_all_strands</span></code>(strands[, copy, join])</p></td>
<td><p>Replace all the strands in the motif with the provided list of strands.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">rotate</span></code>([times])</p></td>
<td><p>Rotate the motif 90 degrees clockwise.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_3d_model</span></code>([filename, config, topology, ...])</p></td>
<td><p>Save the motif 3D structure in a file for structural 3D modeling.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_fasta</span></code>([filename])</p></td>
<td><p>Save the motif sequences in a FASTA file.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">save_text</span></code>([filename_path])</p></td>
<td><p>Save the motif representation as a text file.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">shift</span></code>(shift_vect[, extend])</p></td>
<td><p>Shift the motif of the given shift vector.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">sort</span></code>([key, reverse])</p></td>
<td><p>Sort the strands in the motif.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">strip</span></code>([skip_axis])</p></td>
<td><p>Remove the empty lines/columns in the motif structure.</p></td>
</tr>
<tr class="row-even"><td><p><code class="xref py py-obj docutils literal notranslate"><span class="pre">trace</span></code>(motif_list, pos, direction, limits)</p></td>
<td><p>Trace strands in a motif representation in a recursive manner.</p></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.motif_lib.start_end_stem">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.motif_lib.</span></span><span class="sig-name descname"><span class="pre">start_end_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">up_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'3'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">up_right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'5'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">down_left</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">down_right</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'-'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/motif_lib.html#start_end_stem"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.motif_lib.start_end_stem" title="Link to this definition"></a></dt>
<dd><p>Creates a <cite>Utils</cite> motif representing the start or end of a stem with optional
strand labels. For each position, the acceptable values are:
‘3’, ‘5’, ‘─’, ‘-’, ‘’, or None.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.Utils" title="pyfurnace.design.utils.motif_lib.Utils"><code class="xref py py-class docutils literal notranslate"><span class="pre">Utils</span></code></a></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>up_left</strong><span class="classifier">str or None, optional. Default is ‘3’.</span></dt><dd><p>Label for the top-left strand.</p>
</dd>
<dt><strong>up_right</strong><span class="classifier">str or None, optional</span></dt><dd><p>Label for the top-right strand. Default is ‘5’.</p>
</dd>
<dt><strong>down_left</strong><span class="classifier">str or None, optional</span></dt><dd><p>Label for the bottom-left strand. Default is ‘-‘.</p>
</dd>
<dt><strong>down_right</strong><span class="classifier">str or None, optional</span></dt><dd><p>Label for the bottom-right strand. Default is ‘-‘.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict</span></dt><dd><p>Additional keyword arguments passed to the <cite>Utils</cite> constructor.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Utils</dt><dd><p>An instance of the <cite>Utils</cite> class with the appropriate strands.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.motif_lib.stem_cap_link">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.motif_lib.</span></span><span class="sig-name descname"><span class="pre">stem_cap_link</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/motif_lib.html#stem_cap_link"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.motif_lib.stem_cap_link" title="Link to this definition"></a></dt>
<dd><p>Creates a stem cap motif with a curved connection and vertical segments.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.Utils" title="pyfurnace.design.utils.motif_lib.Utils"><code class="xref py py-class docutils literal notranslate"><span class="pre">Utils</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Utils</dt><dd><p>A stem cap <cite>Utils</cite> object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.motif_lib.vertical_double_link">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.motif_lib.</span></span><span class="sig-name descname"><span class="pre">vertical_double_link</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/motif_lib.html#vertical_double_link"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.motif_lib.vertical_double_link" title="Link to this definition"></a></dt>
<dd><p>Creates a double vertical link motif using two parallel vertical strands.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.Utils" title="pyfurnace.design.utils.motif_lib.Utils"><code class="xref py py-class docutils literal notranslate"><span class="pre">Utils</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Utils</dt><dd><p>A vertical double link <cite>Utils</cite> object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.motif_lib.vertical_link">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.motif_lib.</span></span><span class="sig-name descname"><span class="pre">vertical_link</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/motif_lib.html#vertical_link"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.motif_lib.vertical_link" title="Link to this definition"></a></dt>
<dd><p>Creates a vertical link motif represented by a single vertical strand.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><a class="reference internal" href="#pyfurnace.design.utils.motif_lib.Utils" title="pyfurnace.design.utils.motif_lib.Utils"><code class="xref py py-class docutils literal notranslate"><span class="pre">Utils</span></code></a></span></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>Utils</dt><dd><p>A vertical link <cite>Utils</cite> object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-pyfurnace.design.utils.origami_lib">
<span id="origami-utilities"></span><h3>Origami Utilities<a class="headerlink" href="#module-pyfurnace.design.utils.origami_lib" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.convert_angles_to_dt">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">convert_angles_to_dt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">angles_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#convert_angles_to_dt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.convert_angles_to_dt" title="Link to this definition"></a></dt>
<dd><p>Convert a list of helix angles into corresponding dovetail values based
on a predefined mapping.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">int</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>angles_list</strong><span class="classifier">list of float</span></dt><dd><p>List of helix angles in degrees. Angles will be wrapped modulo 360.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>list of int</dt><dd><p>Corresponding dovetail values for each angle in the input list.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.ipython_clickable_txt">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">ipython_clickable_txt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">origami</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'500'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barriers</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gradient</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">font_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#ipython_clickable_txt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.ipython_clickable_txt" title="Link to this definition"></a></dt>
<dd><p>Generate an interactive, scrollable HTML view of a RNA origami structure
with clickable motifs that display their indexes in a JavaScript alert.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>origami</strong><span class="classifier">Origami</span></dt><dd><p>An Origami object representing the RNA structure.</p>
</dd>
<dt><strong>max_height</strong><span class="classifier">str or int, optional</span></dt><dd><p>The maximum height of the scrollable view in pixels. Default is ‘500’.</p>
</dd>
<dt><strong>barriers</strong><span class="classifier">optional</span></dt><dd><p>Optional barrier data to overlay on the origami representation.</p>
</dd>
<dt><strong>gradient</strong><span class="classifier">bool or str, optional</span></dt><dd><p>Whether to color motifs using a gradient.
If str, interpreted as colormap name from matplotlib.
Default is False.</p>
</dd>
<dt><strong>font_size</strong><span class="classifier">int, optional</span></dt><dd><p>Font size of the text in pixels. Default is 12.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str</dt><dd><p>An HTML string rendered via IPython’s display system.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>This function uses inline CSS and JavaScript for visual styling and interaction.</p></li>
<li><p>Clicking a motif will trigger a JavaScript alert showing its position.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.ipython_display_3D">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">ipython_display_3D</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">origami</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#ipython_display_3D"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.ipython_display_3D" title="Link to this definition"></a></dt>
<dd><p>Display a 3D representation of an Origami structure within a J
upyter notebook using oxDNA.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>origami</strong><span class="classifier">Origami</span></dt><dd><p>The Origami structure to visualize.</p>
</dd>
<dt><strong>**kwargs</strong><span class="classifier">dict, optional</span></dt><dd><p>Additional keyword arguments passed to the <cite>oxdna_conf</cite>
visualization function.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>None</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.ipython_display_txt">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">ipython_display_txt</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">origami_text</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'500'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#ipython_display_txt"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.ipython_display_txt" title="Link to this definition"></a></dt>
<dd><p>Render plain text (e.g., a textual representation of an origami object) as a
scrollable HTML block in Jupyter.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-obj docutils literal notranslate"><span class="pre">None</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>origami_text</strong><span class="classifier">str, Origami</span></dt><dd><p>The content to display in scrollable format.</p>
</dd>
<dt><strong>max_height</strong><span class="classifier">str, optional</span></dt><dd><p>Maximum height of the scrollable box in pixels (default is ‘500’).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>None</dt><dd></dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.simple_origami">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">simple_origami</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dt_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">kl_columns</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">main_stem</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">left_stem_kl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stem_pos</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_terminal_helix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">end_helix_len</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_angles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_start_end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'first'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#simple_origami"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.simple_origami" title="Link to this definition"></a></dt>
<dd><p>Construct an RNA origami object based on a sequence of dovetail values and
kissing loop parameters.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">Origami</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>dt_list</strong><span class="classifier">list of int</span></dt><dd><p>List of dovetail values representing inter-helix connections.</p>
</dd>
<dt><strong>kl_columns</strong><span class="classifier">int, optional</span></dt><dd><p>Number of kissing loop repeats in each helix (default is 1).</p>
</dd>
<dt><strong>main_stem</strong><span class="classifier">int or list of int or list of list of int, optional</span></dt><dd><p>Length(s) of the main stem in each kissing loop.
Can be a single int, a list (same for all loops), or a matrix for per-loop
customization.</p>
</dd>
<dt><strong>left_stem_kl</strong><span class="classifier">int or list of int or list of list of int, optional</span></dt><dd><p>Length(s) of the left stem for each kissing loop. Defaults to automatic
computation.</p>
</dd>
<dt><strong>stem_pos</strong><span class="classifier">int or list of int, optional</span></dt><dd><p>Position(s) of the main stem insertion among helices. Default is 0 for all.</p>
</dd>
<dt><strong>start</strong><span class="classifier">int, optional</span></dt><dd><p>Index of the main stem where origami building starts (default is 0).</p>
</dd>
<dt><strong>add_terminal_helix</strong><span class="classifier">bool, default True</span></dt><dd><p>Whether to prepend and append helices with no dovetails.</p>
</dd>
<dt><strong>end_helix_len</strong><span class="classifier">int, optional</span></dt><dd><p>Length of the stems at the ends of the helices (default is 8).</p>
</dd>
<dt><strong>use_angles</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, interpret <cite>dt_list</cite> as helix angles and convert them to dovetail
values (default is False).</p>
</dd>
<dt><strong>add_start_end</strong><span class="classifier">bool, default True</span></dt><dd><p>Whether to add a start-end motif in the initial helix.</p>
</dd>
<dt><strong>align</strong><span class="classifier">str, optional</span></dt><dd><p>Alignment method for the origami object (default is ‘first’).</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>The assembled Origami structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_2_helix">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_2_helix</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_2_helix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_2_helix" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for a 2-helix structure.
Reference:
1. Krissanaprasit, A. et al. A functional RNA-origami as direct thrombin</p>
<blockquote>
<div><p>inhibitor with fast-acting and specific single-molecule reversal agents in
vivo model. Molecular Therapy 32, 2286-2298 (2024).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the 2-helix RNA structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_3_arms_droplet">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_3_arms_droplet</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_3_arms_droplet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_3_arms_droplet" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for a 3-arms droplet structure.</p>
<p>References:
1. Stewart, J. M. et al. Modular RNA motifs for orthogonal phase separated</p>
<blockquote>
<div><p>compartments. Nat Commun 15, (2024).</p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><dl class="simple">
<dt>Fabrini, G. et al. Co-transcriptional production of programmable RNA condensates</dt><dd><p>and synthetic organelles. Nat. Nanotechnol. 19, 1665-1673 (2024).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Monari, L., Braun, I., Poppleton, E. &amp; Göpfrich, K. PyFuRNAce: An integrated</dt><dd><p>design engine for RNA origami. (2025) doi:10.1101/2025.04.17.647389.</p>
</dd>
</dl>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the 3-arms droplet structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_4_arms_droplet">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_4_arms_droplet</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_4_arms_droplet"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_4_arms_droplet" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for a 4-arms droplet structure.</p>
<p>Reference:
1. Fabrini, G. et al. Co-transcriptional production of programmable RNA condensates</p>
<blockquote>
<div><p>and synthetic organelles. Nat. Nanotechnol. 19, 1665-1673 (2024).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the 4-arms droplet structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_pentagon_tile">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_pentagon_tile</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_pentagon_tile"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_pentagon_tile" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for a pentagon tile structure.
This follows the 3H-4DT design from Geary et al. (2021).</p>
<p>Reference:
1. Geary, C., Grossi, G., McRae, E. K. S., Rothemund, P. W. K. &amp; Andersen, E. S.</p>
<blockquote>
<div><p>RNA origami design tools enable cotranscriptional folding of kilobase-sized
nanoscaffolds. Nat. Chem. 13, 549-558 (2021).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the pentagon tile structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_rectangle_10H_3X">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_rectangle_10H_3X</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_rectangle_10H_3X"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_rectangle_10H_3X" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for a rectangle 10H-3X structure.</p>
<p>References:
1. Monari, L., Braun, I., Poppleton, E. &amp; Göpfrich, K. PyFuRNAce: An integrated</p>
<blockquote>
<div><p>design engine for RNA origami. (2025) doi:10.1101/2025.04.17.647389.</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the rectangle 10H-3X structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_rna_filament">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_rna_filament</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_rna_filament"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_rna_filament" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for an RNA filament structure.</p>
<p>Reference:
1. Tran, M. P. et al. Genetic encoding and expression of RNA origami</p>
<blockquote>
<div><p>cytoskeletons in synthetic cells. Nat. Nanotechnol. 20, 664-671 (2025).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the RNA filament structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.design.utils.origami_lib.template_rna_filament_ispinach">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.design.utils.origami_lib.</span></span><span class="sig-name descname"><span class="pre">template_rna_filament_ispinach</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/design/utils/origami_lib.html#template_rna_filament_ispinach"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.design.utils.origami_lib.template_rna_filament_ispinach" title="Link to this definition"></a></dt>
<dd><p>Generate the RNA origami template for an RNA filament structure.
The tiles contains an ispinach aptamer for fluorescence imaging.</p>
<p>Reference:
1. Tran, M. P. et al. Genetic encoding and expression of RNA origami</p>
<blockquote>
<div><p>cytoskeletons in synthetic cells. Nat. Nanotechnol. 20, 664-671 (2025).</p>
</div></blockquote>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Origami</dt><dd><p>An Origami object representing the RNA filament structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>
<section id="module-pyfurnace.generate.road">
<span id="sequence-generation"></span><h2>Sequence Generation<a class="headerlink" href="#module-pyfurnace.generate.road" title="Link to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.generate.road.generate_road">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.generate.road.</span></span><span class="sig-name descname"><span class="pre">generate_road</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudoknots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'origami'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callback</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zip_directory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">origami_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/generate/road.html#generate_road"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.generate.road.generate_road" title="Link to this definition"></a></dt>
<dd><p>Generate an RNA origami design using the ROAD Revolvr algorithm.</p>
<p>This function orchestrates the structure processing, pseudoknot embedding,
and Perl-based design process for RNA origami, with optional support for
real-time updates and zipped output.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">str</span></dt><dd><p>RNA secondary structure in dot-bracket notation.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str</span></dt><dd><p>Initial sequence to use for the design (must match structure length).</p>
</dd>
<dt><strong>pseudoknots</strong><span class="classifier">str or dict, optional</span></dt><dd><p>Pseudoknot definitions in ROAD-compatible format (string or parsed dict).</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>Base name for the design file (default is ‘origami’).</p>
</dd>
<dt><strong>initial_sequence</strong><span class="classifier">str, optional</span></dt><dd><p>Optional initial sequence to pre-fill as the starting point of the
optimization.</p>
</dd>
<dt><strong>callback</strong><span class="classifier">callable, optional</span></dt><dd><p>A function called during ROAD execution with progress updates:
<cite>callback(structure, sequence, line, stage_name, stage_progress)</cite></p>
</dd>
<dt><strong>verbose</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, prints progress updates to the console.
If <cite>callback</cite> is provided, this is ignored. Default is False.</p>
</dd>
<dt><strong>timeout</strong><span class="classifier">int, optional</span></dt><dd><p>Timeout in seconds for the ROAD optimization
(default is 7200 seconds = 2 hours).</p>
</dd>
<dt><strong>directory</strong><span class="classifier">str, optional</span></dt><dd><p>Directory where files should be written. If None, a temporary directory
is used.</p>
</dd>
<dt><strong>zip_directory</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to generate a <cite>.zip</cite> file with all intermediate and result files.</p>
</dd>
<dt><strong>origami_code</strong><span class="classifier">str, optional</span></dt><dd><p>Source code (e.g., Python) representing the origami design, to be saved
and zipped.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str or tuple</dt><dd><p>If <cite>zip_directory</cite> is False: returns the designed RNA sequence as a string.
If <cite>zip_directory</cite> is True: returns
the tuple <cite>(designed_sequence, path_to_zip_file)</cite>.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>ValueError</dt><dd><p>If the structure and sequence lengths do not match, or if multistrand
‘&amp;’ is used.</p>
</dd>
</dl>
</dd>
</dl>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<ul class="simple">
<li><p>If the working directory doesn’t exist, it is created.</p></li>
<li><p>If the final design file is missing, a warning is issued.</p></li>
</ul>
</div>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.generate.road.parallel_road">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.generate.road.</span></span><span class="sig-name descname"><span class="pre">parallel_road</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">structure</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sequence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pseudoknots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'origami'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">callback</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timeout</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7200</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zip_directory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">origami_code</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial_sequence</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_trials</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_to</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">wait_for_all</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/generate/road.html#parallel_road"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.generate.road.parallel_road" title="Link to this definition"></a></dt>
<dd><p>Run multiple ROAD optimization trials in parallel and return results.</p>
<p>You can choose between returning early on the first successful result
or waiting for all trials to complete. Results can optionally be
ZIP-archived and saved. If the wait_for_all flag is set to True,
the sequences are returned sorted by their MFE frequency in the ensemble
(highest to lowest).</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>structure</strong><span class="classifier">str</span></dt><dd><p>Dot-bracket notation of the target RNA secondary structure.</p>
</dd>
<dt><strong>sequence</strong><span class="classifier">str</span></dt><dd><p>Reference sequence with the same length as the structure.</p>
</dd>
<dt><strong>pseudoknots</strong><span class="classifier">str or dict, optional</span></dt><dd><p>ROAD-style or pyFuRNAce-style pseudoknot definition.</p>
</dd>
<dt><strong>name</strong><span class="classifier">str, optional</span></dt><dd><p>Base name for output files (default: “origami”).</p>
</dd>
<dt><strong>callback</strong><span class="classifier">callable, optional</span></dt><dd><p>Optional function that receives real-time progress updates.</p>
</dd>
<dt><strong>verbose</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, prints progress to console.</p>
</dd>
<dt><strong>timeout</strong><span class="classifier">int, optional</span></dt><dd><p>Timeout in seconds per trial (default: 7200).</p>
</dd>
<dt><strong>zip_directory</strong><span class="classifier">bool, optional</span></dt><dd><p>Whether to save each trial’s output as a ZIP archive.</p>
</dd>
<dt><strong>origami_code</strong><span class="classifier">str, optional</span></dt><dd><p>Python source code to include with the design output.</p>
</dd>
<dt><strong>initial_sequence</strong><span class="classifier">str, optional</span></dt><dd><p>Optional sequence to use as a starting point in optimization.</p>
</dd>
<dt><strong>n_trials</strong><span class="classifier">int, optional</span></dt><dd><p>Number of parallel optimization trials to run.</p>
</dd>
<dt><strong>save_to</strong><span class="classifier">str, optional</span></dt><dd><p>Directory to store ZIP outputs if <cite>zip_directory=True</cite>.</p>
</dd>
<dt><strong>wait_for_all</strong><span class="classifier">bool, optional</span></dt><dd><p>If True, waits for all trials to finish and returns results sorted
by MFE frequency. If False, returns as soon as the first trial succeeds.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>str or Tuple[str, str] or List[str]</dt><dd><ul class="simple">
<li><p>If <cite>wait_for_all=False</cite> returns the first designed sequence.</p></li>
<li><p>If <cite>wait_for_all=True</cite> returns a list of designed sequences.</p></li>
</ul>
</dd>
</dl>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt>RuntimeError</dt><dd><p>If no trials successfully generate a design.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
<section id="prepare">
<h2>Prepare<a class="headerlink" href="#prepare" title="Link to this heading"></a></h2>
<section id="module-pyfurnace.prepare.utils">
<span id="primers"></span><h3>Primers<a class="headerlink" href="#module-pyfurnace.prepare.utils" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.prepare.utils.annealing_temp">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.prepare.utils.</span></span><span class="sig-name descname"><span class="pre">annealing_temp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mts</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_kwargs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'IDT'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/prepare/utils.html#annealing_temp"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.prepare.utils.annealing_temp" title="Link to this definition"></a></dt>
<dd><p>Calculate the annealing temperature for PCR primers using either
the IDT or Phusion method.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>mts</strong><span class="classifier">List[float]</span></dt><dd><p>List of melting temperatures (Tm) for the primers.</p>
</dd>
<dt><strong>seq</strong><span class="classifier">str</span></dt><dd><p>DNA sequence for which the annealing temperature is to be calculated.</p>
</dd>
<dt><strong>tm_kwargs</strong><span class="classifier">dict</span></dt><dd><p>Dictionary of keyword arguments for Tm calculation, such as
salt concentration, DMSO, etc.</p>
</dd>
<dt><strong>method</strong><span class="classifier">str, optional</span></dt><dd><p>Method to use for calculation. Either ‘IDT’ or ‘Phusion’.
Default is ‘IDT’.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>float</dt><dd><p>Calculated annealing temperature, rounded to two decimal places.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>The ‘IDT’ method applies a chemical correction and a weighted average formula.</p></li>
<li><p>The ‘Phusion’ method uses the minimum Tm, with an adjustment for primer length.</p></li>
<li><dl class="simple">
<dt>Certain keys in <cite>tm_kwargs</cite> (e.g., ‘dNTPs’, ‘DMSO’) are handled specifically</dt><dd><p>for the ‘IDT’ method.</p>
</dd>
</dl>
</li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.prepare.utils.auto_design_primers">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.prepare.utils.</span></span><span class="sig-name descname"><span class="pre">auto_design_primers</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_temp=65.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_func=functools.partial(&lt;function</span> <span class="pre">Tm_NN&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nn_table={'init':</span> <span class="pre">(0.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-5.7)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'init_A/T':</span> <span class="pre">(2.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">6.9)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'init_G/C':</span> <span class="pre">(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'init_oneG/C':</span> <span class="pre">(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'init_allA/T':</span> <span class="pre">(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'init_5T/A':</span> <span class="pre">(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'sym':</span> <span class="pre">(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-1.4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'AA/TT':</span> <span class="pre">(-7.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-21.3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'AT/TA':</span> <span class="pre">(-7.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-20.4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'TA/AT':</span> <span class="pre">(-7.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-21.3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'CA/GT':</span> <span class="pre">(-8.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-22.7)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'GT/CA':</span> <span class="pre">(-8.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-22.4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'CT/GA':</span> <span class="pre">(-7.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-21.0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'GA/CT':</span> <span class="pre">(-8.2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-22.2)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'CG/GC':</span> <span class="pre">(-10.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-27.2)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'GC/CG':</span> <span class="pre">(-9.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-24.4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">'GG/CC':</span> <span class="pre">(-8.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">-19.9)}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac1=500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dnac2=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Na=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">K=50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Tris=20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Mg=1.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Method=7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">DMSO</span> <span class="pre">(%)=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mt_method='Nearest</span> <span class="pre">Neighbor'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mt_model='DNA_NN4'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">Primer=500)</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/prepare/utils.html#auto_design_primers"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.prepare.utils.auto_design_primers" title="Link to this definition"></a></dt>
<dd><dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Tuple</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>], <code class="xref py py-class docutils literal notranslate"><span class="pre">List</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code>]]</span></p>
</dd>
</dl>
<dl class="simple">
<dt>Automatically designs forward and reverse primers for a given DNA sequence,</dt><dd><p>optimizing for melting temperature (Tm) and primer quality.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt><strong>seq</strong><span class="classifier">Seq</span></dt><dd><p>The DNA sequence for which primers are to be designed.</p>
</dd>
<dt><strong>target_temp</strong><span class="classifier">float, optional</span></dt><dd><p>The target melting temperature (Tm) for the primers, by default 65.0°C.</p>
</dd>
<dt><strong>tm_func</strong><span class="classifier">Callable, optional</span></dt><dd><p>A function to calculate the melting temperature of a primer sequence,
by default uses <cite>make_tm_calculator()</cite>.</p>
</dd>
</dl>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>final_primers</strong><span class="classifier">list of str</span></dt><dd><p>A list containing the best forward and reverse primer sequences.</p>
</dd>
<dt><strong>final_mts</strong><span class="classifier">list of float</span></dt><dd><p>A list containing the melting temperatures (Tm) of the selected primers.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><dl class="simple">
<dt>The function evaluates primer candidates based on Tm, GC content, length,</dt><dd><p>and dimerization potential.</p>
</dd>
</dl>
</li>
<li><p>Returns empty string and 0 Tm if no suitable primer is found for a direction.</p></li>
</ul>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.prepare.utils.calculate_gc">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.prepare.utils.</span></span><span class="sig-name descname"><span class="pre">calculate_gc</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/prepare/utils.html#calculate_gc"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.prepare.utils.calculate_gc" title="Link to this definition"></a></dt>
<dd><p>Calculate the GC content percentage of a nucleotide sequence.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">float</span></code></span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>seq</strong><span class="classifier">str</span></dt><dd><p>The nucleotide sequence for which to calculate the GC content.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>float</dt><dd><p>The GC content of the sequence as a percentage, rounded to one decimal place.</p>
</dd>
</dl>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Ambiguous nucleotides are ignored in the calculation.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.prepare.utils.check_dimer">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.prepare.utils.</span></span><span class="sig-name descname"><span class="pre">check_dimer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seq1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seq2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dict_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">basepair</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'A':</span> <span class="pre">'T',</span> <span class="pre">'C':</span> <span class="pre">'G',</span> <span class="pre">'G':</span> <span class="pre">'C',</span> <span class="pre">'T':</span> <span class="pre">'A'}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/prepare/utils.html#check_dimer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.prepare.utils.check_dimer" title="Link to this definition"></a></dt>
<dd><p>Check the dimerization of two sequences and return the best dimer found.
The dimerization is checked in an extremely simple way, by aligning the
two sequences and checking for WC basepairing at the same index.
If dict_format is True, return a dictionary with all the dimers found.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Union</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code>]</span></p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.prepare.utils.make_tm_calculator">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.prepare.utils.</span></span><span class="sig-name descname"><span class="pre">make_tm_calculator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">method_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'Nearest</span> <span class="pre">Neighbor'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">model_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DNA_NN4'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">primer_conc</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tm_kwargs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'DMSO</span> <span class="pre">(%)':</span> <span class="pre">0,</span> <span class="pre">'K':</span> <span class="pre">50,</span> <span class="pre">'Method':</span> <span class="pre">7,</span> <span class="pre">'Mg':</span> <span class="pre">1.5,</span> <span class="pre">'Na':</span> <span class="pre">0,</span> <span class="pre">'Primer':</span> <span class="pre">500,</span> <span class="pre">'Tris':</span> <span class="pre">20,</span> <span class="pre">'dNTPs':</span> <span class="pre">0.2,</span> <span class="pre">'mt_method':</span> <span class="pre">'Nearest</span> <span class="pre">Neighbor',</span> <span class="pre">'mt_model':</span> <span class="pre">'DNA_NN4'}</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/prepare/utils.html#make_tm_calculator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.prepare.utils.make_tm_calculator" title="Link to this definition"></a></dt>
<dd><p>Creates a melting temperature (Tm) calculator function for DNA primers using
specified calculation method and model.
Parameters
:rtype: <span class="sphinx_autodoc_typehints-type"><code class="xref py py-class docutils literal notranslate"><span class="pre">callable</span></code></span></p>
<blockquote>
<div><p>Options include “Nearest Neighbor”,
“Empirical GC content”,
and “Wallace, rule of thumb”.
Default is “Nearest Neighbor”.</p>
</div></blockquote>
<dl class="simple">
<dt>model_name<span class="classifier">str, optional</span></dt><dd><p>The nearest neighbor model to use for Tm calculation.
Only relevant if <cite>method_name</cite> is “Nearest Neighbor”.
Default is “DNA_NN4”.</p>
</dd>
<dt>primer_conc<span class="classifier">float, optional</span></dt><dd><p>The concentration of the primer (in nM) used in the calculation.
Only relevant for “Nearest Neighbor” method.
Default is 500.</p>
</dd>
<dt>tm_kwargs<span class="classifier">dict, optional</span></dt><dd><p>Additional keyword arguments for the Tm calculation method.
May include chemical corrections (e.g., DMSO, dNTPs).
Default is a copy of <cite>default_values</cite>.</p>
</dd>
</dl>
<section id="returns">
<h4>Returns<a class="headerlink" href="#returns" title="Link to this heading"></a></h4>
<dl class="simple">
<dt>callable</dt><dd><p>A function that takes a DNA sequence (str) and returns its melting
temperature (float), possibly corrected for chemical additives.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><dl class="simple">
<dt>If DMSO is specified in <cite>tm_kwargs</cite>, the returned function</dt><dd><p>applies a chemical correction to the calculated Tm.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>For “Nearest Neighbor” method, dNTPs and DMSO are removed</dt><dd><p>from the keyword arguments before calculation.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>For “Empirical formulas based on GC content”,</dt><dd><p>DMSO is removed from the keyword arguments before calculation.</p>
</dd>
</dl>
</li>
<li><p>For “Wallace” method, no additional arguments are processed.</p></li>
</ul>
</section>
<section id="examples">
<h4>Examples<a class="headerlink" href="#examples" title="Link to this heading"></a></h4>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tm_calc</span> <span class="o">=</span> <span class="n">make_tm_calculator</span><span class="p">(</span><span class="n">method_name</span><span class="o">=</span><span class="s2">&quot;Nearest Neighbor&quot;</span><span class="p">,</span> <span class="n">primer_conc</span><span class="o">=</span><span class="mi">250</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">tm_calc</span><span class="p">(</span><span class="s2">&quot;ATCGATCG&quot;</span><span class="p">)</span>
<span class="go">62.5</span>
</pre></div>
</div>
</section>
</dd></dl>

</section>
<section id="module-pyfurnace.prepare.oxdna_sim">
<span id="oxdna-simulation-setup"></span><h3>Oxdna simulation setup<a class="headerlink" href="#module-pyfurnace.prepare.oxdna_sim" title="Link to this heading"></a></h3>
<dl class="py function">
<dt class="sig sig-object py" id="pyfurnace.prepare.oxdna_sim.oxdna_simulations">
<span class="sig-prename descclassname"><span class="pre">pyfurnace.prepare.oxdna_sim.</span></span><span class="sig-name descname"><span class="pre">oxdna_simulations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">origami</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">oxdna_directory</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">directory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zip_directory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">temperature</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">37</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mc_relax_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_relax_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_equil_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100000000.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">md_prod_steps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000000000.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pyfurnace/prepare/oxdna_sim.html#oxdna_simulations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pyfurnace.prepare.oxdna_sim.oxdna_simulations" title="Link to this definition"></a></dt>
<dd><p>Prepare the files for the OxDNA simulations. The function saves the
files for oxdna simulations in the specified directory (created if it does not
exist). The function also creates a zip file with the files for the simulations
if the flag zip_directory is set to True. If the directory is set to None, the
program automatically creates a zip file and returns the path to the zip file.</p>
<dl class="field-list simple">
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><span class="sphinx_autodoc_typehints-type"><code class="xref py py-data docutils literal notranslate"><span class="pre">Optional</span></code>[<code class="xref py py-class docutils literal notranslate"><span class="pre">str</span></code>]</span></p>
</dd>
<dt class="field-even">Parameters<span class="colon">:</span></dt>
<dd class="field-even"><dl class="simple">
<dt><strong>origami</strong><span class="classifier">Origami</span></dt><dd><p>The origami object.</p>
</dd>
<dt><strong>oxdna_directory</strong><span class="classifier">str</span></dt><dd><p>The path to the OxDNA executable. It is important to load sequence dependent
parameters.</p>
</dd>
<dt><strong>directory</strong><span class="classifier">str</span></dt><dd><p>The output directory for the MD simulations files. If None, the zip_directory
flag is set to True.</p>
</dd>
<dt><strong>zip_directory</strong><span class="classifier">bool</span></dt><dd><p>If True, the function creates a zip file with the files for the simulations
and returns the path to the zip file. If False, the function returns None.</p>
</dd>
<dt><strong>temperature</strong><span class="classifier">Union[int, float], optional</span></dt><dd><p>The temperature in Celsius, by default 37.</p>
</dd>
<dt><strong>mc_relax_steps</strong><span class="classifier">Union[int, float], optional</span></dt><dd><p>The number of Monte Carlo relaxation steps, by default 5e3.</p>
</dd>
<dt><strong>md_relax_steps</strong><span class="classifier">Union[int, float], optional</span></dt><dd><p>The number of MD relaxation steps, by default 1e7.</p>
</dd>
<dt><strong>md_equil_steps</strong><span class="classifier">Union[int, float], optional</span></dt><dd><p>The number of MD equilibration steps, by default 1e8.</p>
</dd>
<dt><strong>md_prod_steps</strong><span class="classifier">Union[int, float], optional</span></dt><dd><p>The number of MD production steps, by default 1e9.</p>
</dd>
</dl>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><dl class="simple">
<dt>Union[str, None]</dt><dd><p>If zip_directory is True or directory is None, the function returns the path
to the zip file. If zip_directory is False and directory is not None, the
function returns None.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Install and run" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="limitations.html" class="btn btn-neutral float-right" title="Limitations &amp; Workarounds" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Luca Monari.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.


</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

</body>
</html>
